---
title: Report
author: |
 | Project: `r CIDAtools::ProjectName()`
 |
 | Analyst: `r CIDAtools::ProjectAnalyst()`
 |
 | Investigator(s): `r CIDAtools::ProjectPI()`
 |
 | Report generated: `r paste(format(Sys.Date(), '%B %d, %Y'))`
output:
  html_document:
    highlight: espresso
    number_sections: yes
    theme: yeti
    toc: yes
    toc_depth: 2
    toc_float: yes
    code_folding: hide
---

```{r, echo=FALSE, out.width='70%', fig.show='hold'}
logo <- system.file("figures", "sph_cida_wm_blk.png", package="CIDAtools")
knitr::include_graphics(logo)
```

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T, message = F, error = F, warning = F)
library(tidyverse)
library(knitr)
library(kableExtra)
library(table1)
```

# Introduction

The purpose of this report is to identify differentially methylated regions (DMRs) in the naso-epithelial methylation data of vaping and non-vaping adolescents. The previous analysis was a targeted methylation analysis informed by the set of genes found to be differentially expressed in the report "teen_vape_DE_report_yyyy_mm_dd". There was a concern that the burden of multiple testing may still be too large for the relatively small sample size of this study. 

In addition to the burden of small sample size, probes can be auto-correlated across the genome which may further reduce statistical power. To account 

# Methods

Methods text

# Analysis
```{r Data Prep, message=FALSE}
######## DMR Analysis Using ENmix ############
## load up the packages we will need:  (uncomment as required)
require(tidyverse)
require(here)
require(ENmix)
require(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
require(org.Hs.eg.db)
require(gt)
require(gtExtras)
require(data.table)

## Read in the meth results
meth_res <- read_csv(here("DataProcessed/methylation/results/full_res_bacon_2022_12_01.csv")) 

## Access annotation data
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annotation.table <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)


## Convert to Data Frame for easier Access
mapping <- data.frame("Name" = annotation.table$Name,
                      "RefSeq" = annotation.table$UCSC_RefGene_Name,
                      "chrom" = annotation.table$chr,
                      "pos" = annotation.table$pos, 
                      "strand" = annotation.table$strand,
                      "island_name" = annotation.table$Islands_Name,
                      "relation_to_island" = annotation.table$Relation_to_Island)

dim(mapping)

## Merge With EWAS results 
cpg.island.map <- left_join(meth_res, mapping, by = c("CpG_Site" = "Name"))

## Format fields to match bed file 
cpg.island.map <- cpg.island.map %>% 
  mutate(start = pos,
         end = start + 51,
         chrom = gsub("chr", "", chrom),
         chr = factor(chrom, levels = paste0(c(seq(1:22), "X", "Y"))))

## Subset columns needed for Comb-P
dmr.full <- cpg.island.map %>%
  dplyr::rename(p = pval.bacon,
                probe = CpG_Site) %>% 
  dplyr::select(chr, start, end, p) %>% 
  #dplyr::filter(startsWith(probe, "cg")) %>% 
  drop_na(chr) %>% 
  dplyr::arrange(chr, start) %>% 
  dplyr::mutate(chr = factor(chr, levels = c(seq(1:22), "X", "Y"), labels = c(paste0("chr", seq(1,22)), "chrX", "chrY")),
                start = as.integer(start),
                end = as.integer(end)) %>% 
  dplyr::rename(chrom = chr,
                rawp = p)

data.table::fwrite(dmr.full, file = here("DataProcessed/methylation/DMR/bed/dmr_full.bed"),
                   sep = "\t",
                   row.names = F,
                   col.names = F)
```

```{bash Comb-p, eval = F}
##### NEED TO RUN IN Python 2.7 (probably with vitual env) sqlalchemy version = 1.0
# Sort the file using bedtools
FILES="/Users/hawkjona/Documents/CIDA/Sharma_Vaping_Local.nosync/DataProcessed/methylation/DMR/bed/*.bed"

for file in $FILES ; do
  output="${file}_sorted.bed"
  sortBed -i $file > $output
done

# Attatch the appropriate column names for comb-p
FILES="/Users/hawkjona/Documents/CIDA/Sharma_Vaping_Local.nosync/DataProcessed/methylation/DMR/bed/*_sorted.bed"
for file in $FILES ; do
  output="${file}_forCombp.bed"
  echo -e "chrom\tstart\tend\trawp" | cat - $file > $output
done

#Change directory
cd /Users/hawkjona/Documents/CIDA/Sharma_Vaping_Local.nosync

#Activate Virtual Environment with python 2.7.x
source .venv/bin/activate

# Run Comb-p
FILES="/Users/hawkjona/Documents/CIDA/Sharma_Vaping_Local.nosync/DataProcessed/methylation/DMR/bed/*_forCombp.bed"
for file in $FILES ; do
  output="${file}.combp_results"
  comb-p pipeline -c 4 --seed 0.1 --dist 750 -p $output --anno hg19 $file
done

## Moving output files to correct directory
#Set Working Directory
cd /Users/hawkjona/Documents/CIDA/Sharma_Vaping_Local.nosync

mv ./DataProcessed/methylation/DMR/bed/dmr_full.bed_sorted.bed_forCombp.bed.combp_results.anno.hg19.bed ./DataProcessed/methylation/DMR/results/dmr_res_py.bed
```

# Results 

```{r completing_results, message=T}
## Want to read back in results to figure out + and - methylated sites in each DMR

#Read back in the results from comb-p
dmr.res <- read_tsv(here("DataProcessed/methylation/DMR/results/dmr_res_py.bed")) %>% 
  janitor::clean_names() %>% 
  dplyr::rename(chrom = number_chrom)

#Produce key file for dmp res
dmp.key <- cpg.island.map %>% 
  dplyr::select(chrom, pos, CpG_Site, Estimate, pval.bacon) %>%
  drop_na(chrom) %>% 
  mutate(chrom = paste0("chr", chrom))

#Find which cpgs that fall within each DMR and sum up if they are pos or neg methylated
get_meth_dir <- function(dmr, dmp) {
  tmp_res <- matrix(ncol = 3, nrow = dim(dmr)[1])
  colnames(tmp_res) <- c("cpg_sites","n_meth_pos", "n_meth_neg")
  
  for (i in 1:dim(dmr)[1]){
    chr <- dmr$chrom[[i]]
    start <- dmr$start[[i]]
    end <- dmr$end[[i]]
    minp <- dmr$min_p[[i]]
    
    temp <- dmp[dmp$chrom == chr & between(dmp$pos, start, end),]
    cpg_list <- paste(temp$CpG_Site, collapse = ";")
    tmp_res[i,] <- c(cpg_list, sum(temp$Estimate > 0), sum(temp$Estimate < 0))
  }
  
  return(cbind(dmr, tmp_res))
  
}

#Running the function 
dmr.full.anno <- get_meth_dir(dmr.res, dmp.key) %>% 
  mutate(n_meth_pos = as.numeric(n_meth_pos),
         n_meth_neg = as.numeric(n_meth_neg)) %>% 
  arrange(z_sidak_p)

#Subset any rows where number of positively and negatively methylated probes does not sum to n_probes
wrong <- dmr.full.anno[!dmr.full.anno$n_meth_pos + dmr.full.anno$n_meth_neg == dmr.full.anno$n_probes,]

message("There are ", dim(wrong)[1], " DMRs where the number of positively and negatively methylated probes do not sum to n_probes. There p-values are >", round(min(wrong$z_sidak_p), 2), ".")
```


```{r}

dmr.full.sig <- dmr.full.anno %>%
  filter(z_sidak_p < 0.2)

#results table
 dmr.full.sig %>% 
  mutate(sidak.p = if_else(z_sidak_p < 0.001, "<0.001", as.character(round(z_sidak_p, 3)))) %>% 
  dplyr::select(chrom, start, end, sidak.p, n_probes, ref_gene_name, n_meth_pos, n_meth_neg) %>% 
  gt() %>% 
  fmt_integer(columns = c(2, 3)) %>% 
  gt_highlight_rows(rows = which(dmr.full.anno$z_sidak_p < 0.1), 
                    font_weight = "bold", alpha = 0.5) %>% 
  tab_header(title = "Table 1: Top DMRs by Sidak-Adjusted p-value") %>% 
  tab_options(table.width = 700,
              table.font.size = 13)
```


# Conclusions

Interpretations, discussion

# References


# Appendix

<!-- footer -->

---

```{r, echo=FALSE, out.width='70%', fig.show='hold'}
knitr::include_graphics(logo)
```

