---
title: Report
author: |
 | Project: `r CIDAtools::ProjectName()`
 |
 | Analyst: `r CIDAtools::ProjectAnalyst()`
 |
 | Investigator(s): `r CIDAtools::ProjectPI()`
 |
 | Report generated: `r paste(format(Sys.Date(), '%B %d, %Y'))`
output:
  html_document:
    highlight: espresso
    number_sections: yes
    theme: yeti
    toc: yes
    toc_depth: 2
    toc_float: yes
    code_folding: hide
---

```{r, echo=FALSE, out.width='70%', fig.show='hold'}
logo <- system.file("figures", "sph_cida_wm_blk.png", package="CIDAtools")
knitr::include_graphics(logo)
```

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T, message = F, error = F, warning = F)
library(tidyverse)
library(knitr)
library(kableExtra)
library(table1)
```

# Introduction

The purpose of this report is to identify differentially methylated regions (DMRs) in the naso-epithelial methylation data of vaping and non-vaping adolescents. The previous analysis was a targeted methylation analysis informed by the set of genes found to be differentially expressed in the report "teen_vape_DE_report_yyyy_mm_dd". There was a concern that the burden of multiple testing may still be too large for the relatively small sample size of this study. 

In addition to the burden of small sample size, probes can be auto-correlated across the genome which may further reduce statistical power. To account 

# Methods

Methods text

# Analysis
```{r}
######## DMR Analysis Using ENmix ############
## load up the packages we will need:  (uncomment as required)
require(tidyverse)
require(here)
require(ENmix)
require(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
require(org.Hs.eg.db)
require(gt)
require(data.table)

## Read in the meth results
meth_res <- read_csv(here("DataProcessed/methylation/results/full_res_bacon_2022_12_01.csv")) 

## Access annotation data
data("IlluminaHumanMethylationEPICanno.ilm10b4.hg19")
annotation.table <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)


## Convert to Data Frame for easier Access
mapping <- data.frame("Name" = annotation.table$Name,
                      "RefSeq" = annotation.table$UCSC_RefGene_Name,
                      "chrom" = annotation.table$chr,
                      "pos" = annotation.table$pos, 
                      "strand" = annotation.table$strand,
                      "island_name" = annotation.table$Islands_Name,
                      "relation_to_island" = annotation.table$Relation_to_Island)

dim(mapping)

## Merge With EWAS results 
cpg.island.map <- left_join(meth_res, mapping, by = c("CpG_Site" = "Name"))

## Format fields to match bed file 
cpg.island.map <- cpg.island.map %>% 
  mutate(start = pos,
         end = start + 51,
         chrom = gsub("chr", "", chrom),
         chr = factor(chrom, levels = paste0(c(seq(1:22), "X", "Y"))))

## Subset columns needed for Comb-P
dmr.full <- cpg.island.map %>%
  dplyr::rename(p = pval.bacon,
                probe = CpG_Site) %>% 
  dplyr::select(chr, start, end, p) %>% 
  #dplyr::filter(startsWith(probe, "cg")) %>% 
  drop_na(chr) %>% 
  dplyr::arrange(chr, start) %>% 
  dplyr::mutate(chr = factor(chr, levels = c(seq(1:22), "X", "Y"), labels = c(paste0("chr", seq(1,22)), "chrX", "chrY")),
                start = as.integer(start),
                end = as.integer(end)) %>% 
  dplyr::rename(chrom = chr,
                rawp = p)

data.table::fwrite(dmr.full, file = here("DataProcessed/methylation/DMR/bed/dmr_full.bed"),
                   sep = "\t",
                   row.names = F,
                   col.names = F)
```

```{bash}
##### NEED TO RUN IN Python 2.7 (probably with vitual env) sqlalchemy version = 1.0
# Sort the file using bedtools
FILES="/Users/hawkjona/Documents/CIDA/Sharma_Vaping_Local.nosync/DataProcessed/methylation/DMR/bed/*.bed"

for file in $FILES ; do
  output="${file}_sorted.bed"
  sortBed -i $file > $output
done

# Attatch the appropriate column names for comb-p
FILES="/Users/hawkjona/Documents/CIDA/Sharma_Vaping_Local.nosync/DataProcessed/methylation/DMR/bed/*_sorted.bed"
for file in $FILES ; do
  output="${file}_forCombp.bed"
  echo -e "chrom\tstart\tend\trawp" | cat - $file > $output
done

#Change directory
cd /Users/hawkjona/Documents/CIDA/Sharma_Vaping_Local.nosync

#Activate Virtual Environment with python 2.7.x
source .venv/bin/activate

# Run Comb-p
FILES="/Users/hawkjona/Documents/CIDA/Sharma_Vaping_Local.nosync/DataProcessed/methylation/DMR/bed/*_forCombp.bed"
for file in $FILES ; do
  output="${file}.combp_results"
  comb-p pipeline -c 4 --seed 0.1 --dist 750 -p $output --anno hg18 $file
done
```


```{bash}
## Moving files to correct directory
#Set Working Directory
cd /Users/hawkjona/Documents/CIDA/Sharma_Vaping_Local.nosync

mv ./DataProcessed/methylation/DMR/bed/dmr_full.bed_sorted.bed_forCombp.bed.combp_results.anno.hg18.bed ./DataProcessed/methylation/DMR/results/dmr_res_py.bed

```

```{r completing_results}
## Want to read back in results to figure out + and - methylated sites in each DMR

#Read back in the results from comb-p
dmr.res <- read_tsv(here("DataProcessed/methylation/DMR/results/dmr_res_py.bed")) %>% 
  janitor::clean_names() %>% 
  dplyr::rename(chrom = number_chrom) %>% 
  dplyr::mutate(id = seq(1, length(dmr_res_py$chrom), 1), .before = 1)

#Produce key file for dmp res
dmp.key <- cpg.island.map %>% 
  dplyr::select(chrom, pos, CpG_Site, Estimate, pval.bacon) %>%
  drop_na(chrom) %>% 
  mutate(chrom = paste0('chr', chrom)) %>% 
  distinct()

#Find which cpgs 
get_meth_dir <- function(dmr, dmp) {
  tmp_res <- matrix(ncol = 2, nrow = dim(dmr)[1])
  colnames(tmp_res) <- c("n_pos", "n_neg")
  
  for (i in 1:dim(dmr)[1]){
    chr <- dmr$chrom[[i]]
    start <- dmr$start[[i]]
    end <- dmr$end[[i]]
    
    temp <- dmp[dmp$chrom == chr & between(dmp$pos, start, end),]
    tmp_res[i,] <- c(sum(temp$Estimate > 0), sum(temp$Estimate < 0))
  }
  
  return(cbind(dmr, tmp_res))
  
}

test <- get_meth_dir(dmr.res, dmp.key)

```


```{r}
### Create Bedfiles for CpG Islands and DMR Results ###

## Make key file for CpG's to islands
cpg.island.map <- mapping %>% 
  mutate(island_name_clean = gsub(".*:", "", island_name),
         start = sapply(strsplit(island_name_clean, "-"), `[`, 1),
         end = sapply(strsplit(island_name_clean, "-"), `[`, 2))

#Select the columns we want
bedNames <- c("chrom", "start", "end")

#Drop Non-islands
cpg.island.bed <- cpg.island.map[,c(bedNames, "island_name")] %>% 
  tidyr::drop_na(start) %>% 
  distinct()

#Write out .bed file
write_tsv(cpg.island.bed, here("DataProcessed/methylation/DMR/bed/cpg_islands.bed"),
          col_names = FALSE)

## Reformat DMR results to bed file
#Load File
dmr.res <- read_csv(here("DataProcessed/methylation/DMR/results/combp_res.csv"))

## reformat
dmr.bed <- dmr.res %>% 
  mutate("chrom" = paste0("chr", chr)) %>% 
  dplyr::select(bedNames, sidak, nprobe, probe)

## Wirte out dmr as bed file
write_tsv(dmr.bed, here("DataProcessed/methylation/DMR/bed/dmr_res.bed"),
          col_names = FALSE)
```

```{bash}
<!-- ## Intersect DMR and CpG Islands ## -->

<!-- #Change Working Directory -->
<!-- cd /Users/hawkjona/Documents/CIDA/Sharma_Vaping_Local.nosync/DataProcessed/methylation/DMR/bed -->

<!-- # run intersect using bedtools -->
<!-- bedtools intersect -a dmr_res.bed -b cpg_islands.bed -wa -wb > ./dmr_island_overlap.txt -->
```

```{r}
### Read back in overlap resutls ###
## get all the column names back
intersect.colnames <- c(names(dmr.bed), names(cpg.island.bed)) 

dmr_intersect <- read_tsv(here("DataProcessed/methylation/DMR/bed/dmr_island_overlap.txt"), col_names = intersect.colnames) %>% 
  distinct()

dmr_res_py <- read_tsv(here("DataProcessed/methylation/DMR/bed/dmr_res_python.bed")) %>% 
  arrange(z_sidak_p)

write_csv(dmr_res_py, here("DataProcessed/methylation/DMR/results/dmr_res_python.csv"))

#Unlist genes and islands
genelist <- strsplit(mapping$RefSeq, split = ";")
names(genelist) <- mapping$island_name
islands <- unlist(genelist)
islands <- islands[names(islands) != ""]



#Create Key file for islands to genes
islands.genes.map <- data.frame(island_name = names(islands),
                                gene_name = islands) %>% 
  distinct() %>% 
  group_by(island_name) %>%
  summarize(gene_name = paste(unique(gene_name), collapse = "; ")) %>%
  ungroup()

#Join key with intersect file
dmr_intersect <- left_join(dmr_intersect, islands.genes.map, by = "island_name")

# Drop unneeded columns
dmr_intersect <- dmr_intersect %>% 
  dplyr::select(-c(chrom...7, start...8, end...9))

# Rename columns
names(dmr_intersect) <- c(bedNames, "sidak", "nprobe", "probe", "island_name", "gene_name")

# Create an ID filed in order to rejoin the number of positively and negatively methylated probes
dmr_intersect$id <- seq(1,length(dmr_intersect$chrom), 1)

# Prepare probes to lookup in methylation results
probe.map <- dmr_intersect %>% 
  dplyr::select(id, probe) %>% 
  separate_rows(probe, sep = ";")

# Join for lookup
probe.map.full <- left_join(probe.map, meth_res, by = c("probe" = "CpG_Site"))

#Create a way to track positive and negative methylation
probe.map.full$meth_dir <- if_else(probe.map.full$Estimate > 0, "+", "-")

#Create summary of positive and negative probes
probe.dir.match <- probe.map.full %>% 
  group_by(id) %>% 
  summarise(n_meth_pos = sum(meth_dir == "+"),
            n_meth_neg = sum(meth_dir == "-")) %>% 
  ungroup()

# Join back to original results
dmr.full.res <- left_join(dmr_intersect, probe.dir.match, by = "id") %>% 
  dplyr::select(chrom, start, end, sidak, nprobe, n_meth_pos, n_meth_neg, island_name, gene_name) %>% 
  arrange(sidak)

#Write out the Results
#write_csv(dmr.full.res, here("DataProcessed/methylation/DMR/results/combp_format_res.csv"))


  
```

# Results 
```{r}
#results table
dmr.full.res %>% 
  filter(sidak < 0.2) %>% 
  mutate(sidak.p = if_else(sidak < 0.001, "<0.001", as.character(round(sidak, 3)))) %>% 
  dplyr::select(chrom, start, end, sidak.p, nprobe, n_meth_pos, n_meth_neg, island_name, gene_name) %>% 
  gt() %>% 
  fmt_integer(columns = c(2, 3)) %>% 
  tab_header(title = "Table 1: Top DMRs by Sidak-Adjusted p-value") %>% 
  tab_options(table.width = 700,
              table.font.size = 13)
```


# Conclusions

Interpretations, discussion

# References


# Appendix

<!-- footer -->

---

```{r, echo=FALSE, out.width='70%', fig.show='hold'}
knitr::include_graphics(logo)
```

