---
title: Targeted Methylation Results
author: |
 | Project: Differential Gene Expression and Pulmonary Function in Vaping LatinX Adolescents
 |
 | Analyst: Trent Hawkins
 |
 | Investigator(s): Sunita Sharma, MD
 |
 | Report generated: `r paste(format(Sys.Date(), '%B %d, %Y'))`
output:
  html_document:
    highlight: espresso
    number_sections: no
    theme: yeti
    toc: yes
    toc_depth: 2
    toc_float: yes
csl: ama.csl
bibliography: references.bib
---

```{r, echo=FALSE, out.width='70%', fig.show='hold'}
logo <- system.file("figures", "sph_cida_wm_blk.png", package="CIDAtools")
knitr::include_graphics(logo)
```

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = F, warning = F, error = F)
library(tidyverse)
library(knitr)
library(kableExtra)
library(table1)
require(here)
library(gt)


kablize <- function(tab, digits = 3) {
  kable(tab,digits = digits, booktabs = T) %>% 
  kableExtra::kable_styling(latex_options = c("hold_position", "striped"), position = "center")
}

format_num <- function(number, digits = 0, format = "f") {
  formatC(number, digits = digits, format = format, big.mark = ",")
}

# Load the results from code "05_targeted_meth.R"
sigCites.final <- read_csv(here("DataProcessed/methylation/Targeted/results/results_targeted_2023_01_19.csv"))

# Load RNASeq REsults
RNAseq.res <- read_csv(here("DataProcessed/rna_seq/differential_expression/full_analysis/full_vape_res_2022_10_13.csv"))

RNAseq.sig <- RNAseq.res %>% 
  dplyr::filter(padj < 0.05)
```

# Introduction

The purpose of this report is to discover differential methylation among the subset of CpG sites targeted by the genes found to be differentially expressed in the report "teen_vape_DE_report_*yyyy_mm_dd*". Since this may be considered an *exploratory analysis*, a seperate targeted analysis which includes only genes annotated to some pertinent Gene Ontologies will be presented also. Due to the limited sample size, these targeted analyses were chosen to increase statistical power by limiting the effects of multiple testing.  

# Methods

In order to create a targeted subset of CpG Sites, we first mapped CpG sites to their respective genes. This task was accomplished using annotation for Illumina's EPIC methylation arrays @kasper2017. After which, a subset of CpG Sites was selected based upon the 7,136 genes found to be significantly associated with whether the subject had vaped in the last six months in a previous analysis (FDR < 0.05). Finally, we repeated the False Discovery Rate (FDR) p-value adjustment for multiple testing using the bacon-corrected p-values presented in the previous EWAS report ("DMP_Results_*yyyy_mm_dd*").

To focus the analysis further, we conducted another targeted analysis which included only the CpG sites which were annotated to the genes found in the Gene Ontologies [@dolinski2000gene, @gene2021gene] named "inflammatory response", "immune system process", "immune response", and "cilium". This search also included CpG sites annotated to the gene "AHRR", which has been the subject of tobacco-related research @richmond2021investigating. We again repeated the FDR p-value adjustment for only the CpG sites associated with these genes. 

# Results 

Results are presented below for both the *full* targeted methylation analysis and the *restricted* targeted analysis. The *full* targeted analysis included all CpG sites which mapped to differentially expressed genes from previous analyses. The *restricted* analysis was limited to CpG sites which mapped only to those genes included in Gene Ontologies "inflammatory response", "immune system process", "immune response", and "cilium" in addition to those which mapped to the gene "AHRR". 

## Full Targeted Analysis 

```{r}
sigCites.generelate <- left_join(sigCites.final, RNAseq.res, by = c("symbol" = "gene_name"))

sigCite <- sigCites.generelate %>% 
  filter(fdr.target < 0.2) %>% 
  dplyr::arrange(fdr.target, Estimate)

sigGenes <- RNAseq.res %>% 
  filter(gene_name %in% sigCite$symbol)

diffmeth.genes <- sigCites.generelate %>% 
  filter(fdr.target < 0.1) %>% 
  dplyr::arrange(fdr.target, Estimate)
```


`r format_num(nrow(RNAseq.sig))` differentially expressed genes mapped to `r format_num(length(unique(sigCites.final$cpg)))` unique CpG sites. After subsetting those sites from the full EWAS, only one was significantly differentially methylated using a type-I error rate of 0.05. Increasing the allowed type-I error rate to 0.1 allowed for one additional site. At a type-I error rate of 0.2, there were `r nrow(sigCite)` differentially methylated positions. Those `r nrow(sigCite)` positions mapped to `r nrow(sigGenes)` unique genes. Table 1 gives the top 10 sorted by FDR adjusted p-value. 

```{r}

sigCite %>% 
  arrange(fdr.target, Estimate) %>% 
  mutate(pval.bacon = if_else(pval.bacon < 0.001, "< 0.001", as.character(pval.bacon)),
         gene.reg = if_else(log2FoldChange > 0, '+', '-')) %>% 
  dplyr::select(cpg, symbol, Estimate, `Std. Error`, `t value`, pval.bacon, fdr.target, gene.reg) %>% 
  dplyr::rename("CpG" = cpg,
                "Associated Gene Name" = symbol,
                "Estimate (M)" = Estimate,
                "p-value" = pval.bacon,
                "FDR" = fdr.target,
                "Associated Gene Reg." = gene.reg) %>%
    .[1:10,] %>% 
  gt() %>% 
  fmt_number(columns = c(3, 4, 5, 7),
             decimals = 3) %>% 
  tab_header(title = "Table 1: Top 10 Targeted CpG Sites by p-value") %>% 
  tab_options(table.width = 700,
              table.font.size = 13)



```



Overall, there was only one CpG site in the targeted analysis that may be considered significantly differentially expressed at a type-I error rate of 0.05. That site is `r sigCite$cpg[1]` and it maps to the gene `r sigCite$symbol[1]` using Illumina's EPIC annotation. In the context of this analysis, the site is *less* methylated in vapers compared to non-vapers. This result is consistent with RNASeq results where the gene was found to be up-regulated (`r format_num(diffmeth.genes$log2FoldChange[1], digits = 3)`, p = `r format_num(diffmeth.genes$padj[1], digits = 2)`). One additional site, `r sigCite$cpg[2]` may be considered marginally significant at an FDR-adjusted p-value of `r format_num(sigCite$fdr.target[2], digits = 2)`. That position maps to the gene `r sigCite$symbol[2]`. That gene was also found to be up-regulated in RNASeq results (`r format_num(diffmeth.genes$log2FoldChange[2], digits = 3)`, p = `r format_num(diffmeth.genes$padj[2], digits = 2)`). **Figure 1** displays boxplots of the estimated relative methylation comparing vapers to non-vapers. 

```{r}
top5 <- sigCite %>% 
  arrange(fdr.target, Estimate) %>% 
  .$cpg %>% 
  .[1:5]

gene.map <- sigCite %>% 
  dplyr::select(cpg, symbol)

mvals <- read_tsv(here("DataProcessed/methylation/QC/methylation_mvals_final_2022_09_27.txt")) 

mvals.fullTarget <- mvals %>% 
  dplyr::filter(CpG_Site %in% top5) %>%
  column_to_rownames(var = "CpG_Site") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "sentrix_name")
  
clin_metadata <- read_csv(here("DataProcessed/clinical_metadata/master_clinical_metadata_2022_09_02.csv")) 

mvals_plot <- left_join(mvals.fullTarget, clin_metadata, by = "sentrix_name")%>% 
  drop_na(vape_6mo_lab, methylation_id) %>% 
  pivot_longer(cols = 2:6, names_to = "cpg", values_to = "estimate") %>% 
  left_join(., gene.map, by = "cpg") %>% 
  dplyr::mutate(label = paste0(cpg, " (", symbol, ")"))


vape_cols <- RColorBrewer::brewer.pal(3, "Dark2")[1:2]

mvals_plot %>%
  ggplot(aes(x = vape_6mo_lab, y = estimate, fill = vape_6mo_lab)) +
  geom_boxplot() +
  facet_wrap(~label) +
  scale_fill_manual(values = vape_cols) +
  theme(legend.position = "bottom",
        axis.text.x = element_blank()) +
  labs(fill = "Vape Status (6 mo)",
       title = "Figure 1: M-values displayed as 'CpG Site (Associated Gene)'",
       y = "Estimate (M)",
       x = element_blank())
```

## Restricted Targeted Analysis 
```{r}
require(tidyverse)
require(here)
require(org.Hs.eg.db)
require(minfi)
require(readxl)
require(missMethyl)
require(clusterProfiler)
require(openxlsx)


# Needed function from missMethyl source code -----------------------------

.getFlatAnnotation <- function(array.type=c("450K","EPIC"),anno=NULL)
  # flatten 450k or EPIC array annotation
  # Jovana Maksimovic
  # 18 September 2018
  # Updated 18 September 2018
  # Modified version of Belida Phipson's .flattenAnn code
{
  if(is.null(anno)){
    if(array.type=="450K"){
      anno <- minfi::getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19::IlluminaHumanMethylation450kanno.ilmn12.hg19)
    } else {
      anno <- minfi::getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19::IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
    }
  }
  
  # get rid of the non-CpG sites
  ann.keep<-anno[grepl("^cg",anno$Name),]
  
  # get rid of CpGs that are not annotated
  missing<-ann.keep$UCSC_RefGene_Name==""
  ann.keep<-ann.keep[!missing,]
  
  # get individual gene names for each CpG
  geneslist<-strsplit(ann.keep$UCSC_RefGene_Name,split=";")
  names(geneslist)<-rownames(ann.keep)
  
  grouplist<-strsplit(ann.keep$UCSC_RefGene_Group,split=";")
  names(grouplist)<-rownames(ann.keep)
  
  flat<-data.frame(symbol=unlist(geneslist),group=unlist(grouplist))
  flat$symbol<-as.character(flat$symbol)
  flat$group <- as.character(flat$group)
  
  flat$cpg<- substr(rownames(flat),1,10)
  
  #flat$cpg <- rownames(flat)
  flat$alias <- suppressWarnings(limma::alias2SymbolTable(flat$symbol))
  
  #eg <- toTable(org.Hs.egSYMBOL2EG)
  eg <- suppressMessages(AnnotationDbi::select(org.Hs.eg.db, 
                                               keys=AnnotationDbi::keys(org.Hs.eg.db), 
                                               columns=c("ENTREZID","SYMBOL"), 
                                               keytype="ENTREZID"))
  colnames(eg) <- c("gene_id","symbol")
  
  m <- match(flat$alias,eg$symbol)
  flat$entrezid <- eg$gene_id[m]
  flat <- flat[!is.na(flat$entrezid),]
  return(flat)
}

# Read in RNASeq Results, Map to ENTREZ, and extract sig genes --------------------------------------------------

## Read in results
RNAseq.res <- read_csv(here("DataProcessed/rna_seq/differential_expression/full_analysis/full_vape_res_2022_10_13.csv"))

GO.res <- read_csv(here("DataProcessed/gsea/go_gene_level_res_2023_01_19.csv"))

## Pathways of interest
pathways <- c("inflammatory response", "immune system process", "immune response", "cilium")

## Subset GO genes of interest (and AHRR gene)
pathway.res <- GO.res %>% 
  dplyr::filter(Pathway %in% pathways)

## Get significant genes
GO.sig <- pathway.res %>% 
  filter(gene_level_fdr < 0.05)


# Read Methylation Results and extract CpG Sites mapped to sig Genes --------

## Read in the meth results
meth_res <- read_csv(here("DataProcessed/methylation/EWAS/results/full_res_bacon_2022_12_01.csv"))

## Genes to CpG Key
genes_to_cpg <- .getFlatAnnotation(array.type = "EPIC")

## Filter for CpGs associated with significant genes
sigGene.map <- genes_to_cpg %>% 
  dplyr::filter(symbol %in% GO.sig$gene_name | symbol == "AHRR")

## Get unique CpG loci
sigGene.cpgs <- unique(sigGene.map$cpg)

## Subset sig loci from meth results
methRes.target <- meth_res %>% 
  dplyr::filter(CpG_Site %in% sigGene.cpgs)

## Re-adjust the p-values
methRes.target$fdr.target <- p.adjust(methRes.target$pval.bacon, method = "fdr")

## Subset new sig cpgs
sig.target.cpgs <- methRes.target %>% 
  dplyr::filter(fdr.target < 0.2)

## Create dataframe for final results
sigCites.final <- left_join(sigGene.map, methRes.target %>% dplyr::select(CpG_Site, Estimate, `Std. Error`, `t value`, pval.bacon, fdr.target), by = c("cpg" = "CpG_Site")) %>% 
  drop_na(fdr.target) %>% 
  dplyr::select(cpg, symbol, Estimate, `Std. Error`, `t value`, pval.bacon, fdr.target) %>% 
  distinct()

## Only results we want to report to the investigator
sigCites.report <- sigCites.final %>%
  left_join(., RNAseq.res, by = c("symbol" = "gene_name")) %>% 
  dplyr::mutate(gene.reg = if_else(log2FoldChange > 0, '+', '-')) %>% 
  filter(fdr.target < 0.2) %>% 
  dplyr::arrange(fdr.target) %>% 
  distinct() %>% 
  dplyr::select(cpg, symbol, Estimate, `Std. Error`, `t value`, pval.bacon, fdr.target, gene.reg) %>% 
  dplyr::rename("CpG Site" = cpg,
                "Symbol" = symbol,
                "Estimate (M)" = Estimate,
                "pval" = pval.bacon,
                "FDR" = fdr.target)

## Get records for related genes
sigCites.generelate <- RNAseq.res %>% 
  dplyr::filter(gene_name %in% sigCites.report$Symbol) %>% 
  dplyr::rename("EnsemblID" = ensg,
                "Symbol" = gene_name,
                "Std. Error" = lfcSE,
                "t-value" = stat,
                "pval" = pvalue,
                "FDR" = padj) %>% 
  dplyr::select(-c(gene_type, baseMean))

```

There were `r length(unique(GO.sig$gene_name))` differentially expressed genes found in the Gene Ontologies "inflammatory response", "immune system process", "immune response", and "cilium". Those `r length(unique(GO.sig$gene_name))` genes and "AHRR" mapped to `r length(unique(sigGene.map$cpg))` unique CpG sites. After correcting those subsetted sites using FDR, `r length(sigCites.report$Symbol)` sites were found to be differentially methylated at a type-I error rate of 0.2. Figure 2 shows boxplots of relative methylation among the two loci associated with the gene 'AHRR' which have been previously shown to serve as biomarkers for tobacco smoke exposure @richmond2021investigating.

```{r}
# Sites from lit review that are typically demethylated from tobacco smoke exposure
tobacco.sites <- c("cg05575921", "cg21161138")

#Subset those sites from the results
tobacco.res <- sigCites.final %>% 
  dplyr::filter(cpg %in% tobacco.sites)

mvals.tobacco <- mvals %>% 
  dplyr::filter(CpG_Site %in% tobacco.sites) %>%
  column_to_rownames(var = "CpG_Site") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "sentrix_name")
  
clin_metadata <- read_csv(here("DataProcessed/clinical_metadata/master_clinical_metadata_2022_09_02.csv")) 

mvals_plot <- left_join(mvals.tobacco, clin_metadata, by = "sentrix_name")%>% 
  drop_na(vape_6mo_lab, methylation_id) %>% 
  pivot_longer(cols = 2:3, names_to = "cpg", values_to = "estimate") %>% 
  left_join(., gene.map, by = "cpg") %>% 
  dplyr::mutate(label = paste0(cpg, " (AHRR)"))


vape_cols <- RColorBrewer::brewer.pal(3, "Dark2")[1:2]

mvals_plot %>%
  ggplot(aes(x = vape_6mo_lab, y = estimate, fill = vape_6mo_lab)) +
  geom_boxplot() +
  facet_wrap(~label) +
  scale_fill_manual(values = vape_cols) +
  theme(legend.position = "bottom",
        axis.text.x = element_blank()) +
  labs(fill = "Vape Status (6 mo)",
       title = "Figure 2: M-values boxplots for 'AHRR' CpG sites of interest",
       y = "Estimate (M)",
       x = element_blank())
```

**Figure 2** shows little noticeable difference in methylation at the sites `r tobacco.sites[1]` or `r tobacco.sites[2]`. 

# Conclusions
  In the case of this analysis, the full targeted analysis showed a limited number of methylation loci that displayed differential methylation between vapers and non-vapers. The relevance of these sites to the mechanisms behind vaping are yet to be determined. Limiting the targeted analysis only to the sites associated with relevant Gene Ontologies as determined by the principal investigators revealed no additional differential methylation patterns between vapers and non-vapers. The next analysis will examine the epigenome for differentially methylated regions, which will further reduce the burden of multiple testing and highlight methylation patterns among vapers and non-vapers.  

# References
::: {#refs}
:::
---

```{r, echo=FALSE, out.width='70%', fig.show='hold'}
knitr::include_graphics(logo)
```

