---
title: Differential Expression Report
author: |
 | Project: `r CIDAtools::ProjectName()`
 |
 | Analyst: `r CIDAtools::ProjectAnalyst()`
 |
 | Investigator(s): `r CIDAtools::ProjectPI()`
 |
 | Report generated: `r paste(format(Sys.Date(), '%B %d, %Y'))`
output:
  html_document:
    highlight: espresso
    number_sections: no
    theme: yeti
    toc: yes
    toc_depth: 2
    toc_float: yes
    code_folding: hide
---

```{r, echo=FALSE, out.width='70%', fig.show='hold'}
logo <- system.file("figures", "sph_cida_wm_blk.png", package="CIDAtools")
knitr::include_graphics(logo)
knitr::opts_chunk$set(warning = F, message = F, error = F, echo = T, include = T)
```

------------------------------------------------------------------------

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(RColorBrewer)
library(tidyverse)
library(knitr)
library(kableExtra)
library(arsenal)
library(here)
library(readr)
library(coin)
library(latex2exp)
library(ggpubr)
library(DESeq2)
library(RUVSeq)
library(janitor)
library(dendextend)
library(WGCNA)
library(minfi)

kablize <- function(tab, digits = 3) {
  kable(tab,digits = digits, booktabs = T) %>% 
  kableExtra::kable_styling(latex_options = c("hold_position", "striped"), position = "center")
}

format_num <- function(number, digits = 0, format = "f") {
  formatC(number, digits = digits, format = format, big.mark = ",")
}
```

```{r import data, include=FALSE}
#import data
tab1_dat <- read_csv(here("DataProcessed/clinical_metadata/master_clinical_metadata_2022_09_02.csv"))

#filter out 1 NA for vape status
tab1_dat <- tab1_dat %>% 
  drop_na(vape_6mo_lab, rna_id) %>% 
  filter(rna_id != "Sample35")
  

```

# Introduction

The purpose of this report is to present results from differential expression analysis for Dr. Sunita Sharma's study assessing the effects of vaping on pulmonary function in adolescents. Additional background and the pre-processing steps taken can be found in the 'Metadata Exploratory Report' included in this series of statistical reports. 

# Methods

*All analyses performed using `r R.version$version.string`*

### Clinical Data Processing

Vape Status is self-reported and defined by vaping within the 6-month period leading up to the survey. In order to obtain the vape status and recruitment center metrics, responses were aggregated over several survey questions. This process is more clearly described in the 'Exploratory Report'. 

### Gene-Count Processing

*Annotation: Ensembl annotation for GrCH 39 ver. 37*\
*Removal of Unwanted Variance: edgeR `r package.version("edgeR")` and RUVSeq `r package.version("RUVSeq")`*\
*Differential Expression Analysis: DESeq2 `r package.version("DESeq2")`*\

Gene filtering and RUVr normalization were previously conducted for the exploratory analysis. An abbreviated version of the results will be presented here, but further detail can be found in the exploratory report. The model used to obtain the residuals for RUVr included *vape status*, *sex*, and *age* as model covariates.

Notice that recruitment center was not incorporated into this model, as we are *not* interested in keeping any variability based on recruitment center in the data set. 

### Model Fitting
The differential expression analysis will use RUVr with k = 2 factors. Gene counts were fit using negative binomial models from the R package DESeq2. To account for multiple testing the False Discovery Rate (FDR) was calculated, and significance set at FDR < 0.05.

To parse out the effects that vape status and recruitment center have on gene expression, the following models were fit and compared using Likelihood Ratio Tests (LRTs) in DESeq2 \:

Full Model: $raw \ read \ count \sim vape \ status + recruitment \ center + sex + age + ruv_1 + ruv_2$  

Reduced Model (testing vape status and recruitment center): $raw \ read \ count \sim \  sex + age + ruv_1 + ruv_2$  

Reduced Model (testing vape status): $raw \ read \ count \sim recruitment \ center + sex + age + ruv_1 + ruv_2$  

Reduced Model (testing recruitment center): $raw \ read \ count \sim vape \ status + sex + age + ruv_1 + ruv_2$  

Likelihood Ratio Tests (LRTs) were used to compare full and reduced models. All LRTs used the same full model, which included vape status (binary Y/N) and recruitment center (3-level categorical), sex (binary M/F), age (continuous), and two normalization factors from the RUVr Analysis. In each scenario above, the LRT examines if the additional covariates included in the full model and not the reduced model significantly increase model fit. The reduced model for vape status and recruitment center tests if vape status *and* recruitment center (together) significantly explain gene expression after adjusting for sex, age, and the two RUV factors. The other models test similarly, but only for one covariate at a time. The model we are most interested in is the reduced model testing only vape status. The others were included to investigate the presence of recruitment center bias in the data.

# Results  

After removing participants with missing RNA-Seq data or missing values for vape status, we are left with **n = 50** subjects. Previous analyses showed **n = 12** participants had vaped in the last 6 months. This analysis will use **n = 13** participants who had vaped in the last 6 months.

### Table 1: Subject Demographics

```{r Table 1, results='asis'}

#create table 1
vape_6mo_table1 <- tab1_dat %>% 
  tableby(vape_6mo_lab ~ sex_lab + age + recruitment_center,
          data = ., digits = 1, test = FALSE )

#Fix Labels
arsenal::labels(vape_6mo_table1) <- c(vape_6mo_lab = "Vaped in last 6 months", 
                             age = "Age (yrs)", sex_lab = "Sex", 
                             recruitment_center = "Recruitment Center")

#Print Tables
summary(vape_6mo_table1, pfootnote = T)
```

```{r}
#Figures
#By sex (n = 50)
sex_plot_trial <- tab1_dat %>% 
  group_by(sex_lab, vape_6mo_lab) %>% 
  summarise(N = n())

bar_sex <- sex_plot_trial %>% 
  ggplot(aes(x = sex_lab, y = N, fill = vape_6mo_lab)) + 
  geom_bar(stat = 'identity', position = 'dodge') +
  labs(y = "Count", fill = "Vape Status (6 mo)") + 
  ggtitle("Sex") +
  scale_fill_manual(values = brewer.pal(3, "Set2"))+
  theme(axis.title.x=element_blank())

#By Latino(n = 50)
bar_latino <- tab1_dat %>% 
  ggplot(aes(x = latino_lab, fill = vape_6mo_lab))+
  geom_bar(position = "dodge")+
  labs( y = "Count", fill = "Vape Status (6 mo)") +
  ggtitle("LatinX") +
  scale_fill_manual(values = brewer.pal(3, "Set2")) +
  theme(axis.title.x=element_blank())

#Fix Recruitment Center Bar Plot
center_plot_trial <- tab1_dat %>% 
  group_by(recruitment_center, vape_6mo_lab) %>% 
  summarise(N = n())

center_plot_trial[nrow(center_plot_trial) + 1,] <- NA

center_plot_trial[6,] <- list("Aurora", "Vaped in Last 6 Months", 0)

#By Recruitment Center
recruitment_center_hist <- center_plot_trial %>% 
  ggplot(aes(x = recruitment_center, y = N, fill = vape_6mo_lab)) + 
  geom_bar(stat = 'identity', position = 'dodge') +
  labs(y = "Count", fill = "Vape Status (6 mo)") + 
  ggtitle("Center")+
  scale_fill_manual(values = brewer.pal(3, "Set2")) +
  theme(axis.title.x=element_blank())

```

### Figure 1: Select Subject Demographics by Vape Status

Most vaping subjects were recruited in Pueblo (91%) and identified as LatinX (85%). 53% of subjects were female as confirmed from the available methylation data.

```{r Demographic Figure, fig.height=6, fig.width=9}
#Demographic Figures
ggarrange(recruitment_center_hist, ggarrange(bar_sex, bar_latino, ncol = 2, nrow = 1, legend = "none"), nrow = 2, common.legend = T, legend = "bottom")
```

Due to the demographic imbalances present in the dataset, it is important to look into potential sources of multicollinearity or confounding. Figure 2 characterizes some of the relationships present in the data using Fischer's Exact tests.  

### Figure 2: Correlation of Model Predictors  

```{r}
source(here("Code/01_variable_correlation.R"), local = knitr::knit_global())

correlation_matrix_plot
```

There is a significant correlation between vape status and recruitment center (p < 0.001). The t-test for vape status and age (p = 0.522) show no significant correlation.  

*t-test: Vape Status and Age*
```{r}
age_report_tab
```


## Gene-Count Processing  

There were **n = 47** samples present with both demographic and RNA-Seq data (see the Exploratory Report for further details). There were 60,651 genes before filtering out poor-quality genes. This analysis will include a total of **`r format_num(length(filtered_gene_count[,1]), 0)`** genes. 

### Figure 3: Relative Log Expression and Principal Component Analysis for RUV

The Relative Log Expression (RLE) and Principal Component Analysis (PCA) of read counts for each sample both before and after transformation using RUVr with k = 2 factors. There was an initial concern for Sample 12 as an outlier.
  
```{r fig.height=7, fig.width= 10}
source(here("Code/03c_gene_filter_and_ruv_CLEAN_for_DE.R"), local = knitr::knit_global())

#Make nicer plot
par(mfrow=c(2,1), cex = .75, cex.lab = 1.5, mar=c(2,5,0.5,0))

#Look at raw sample
colors <- brewer.pal(3,"Set2")
par(mfrow = c(2,1))

plotRLE(ruv_prep, outline=FALSE, ylim=c(-4, 4), col = colors[metadata_joined$vape_6mo_lab], ylab = "No Transformation")
legend(x = "bottomleft", legend = c("Vaped", "Did Not Vape"), fill = c(colors[2], colors[1]))

plotRLE(ruv_k2, outline=FALSE, ylim=c(-2,2), col = colors[metadata_joined$vape_6mo_lab], ylab = "RUVr (k = 2)")
abline(h = c(.5, -.5), col = c("red", "red"))
legend(x = "bottomright", legend = c("Vaped", "Did Not Vape"), fill = c(colors[2], colors[1]))

```

```{r fig.height=4, fig.width= 10}
#Make nicer plot
par(mfrow=c(1,2), cex = .75)

#No Transformation
plotPCA(ruv_prep, col = colors[metadata_joined$vape_6mo_lab], main = "No Transformation", xlim = c(-0.7, 0.7))
legend(x = "topright", legend = c("Vaped", "Did Not Vape"), fill = c(colors[2], colors[1]))

#k = 2
plotPCA(ruv_k2, col = colors[metadata_joined$vape_6mo_lab], main = "RUVr (k = 2)", xlim = c(-0.7, 0.7))
legend(x = "topright", legend = c("Vaped", "Did Not Vape"), fill = c(colors[2], colors[1]))
```  

Using two normalization factors, Sample 12 is no longer a suspected outlier and overall, two factors sufficiently normalize the gene expression counts.

## Differential Expression Analysis

### Figure 4: P-value Distributions (Unadjusted)
```{r fig.height= 4, fig.width = 10}
######################### Read in Results ###############################
vape_center_res <- read_csv(here("DataProcessed/rna_seq/differential_expression/full_analysis/vape_center_res_2022_10_13.csv"))
vape_res <- read_csv(here("DataProcessed/rna_seq/differential_expression/full_analysis/full_vape_res_2022_10_13.csv"))
center_res <- read_csv(here("DataProcessed/rna_seq/differential_expression/full_analysis/center_res_2022_10_13.csv")) %>% 
  dplyr::mutate(padj = p.adjust(pvalue, method = "fdr"))

#Read in log-normalized counts (See 04_DESeq2_DE_Model_fit.R)
vape_center_counts <- read_csv(here("DataProcessed/rna_seq/differential_expression/full_analysis/vape_center_ruv_normcounts_2022_10_13.csv"))
vape_counts <- read_csv(here("DataProcessed/rna_seq/differential_expression/full_analysis/vape_ruv_normcounts_2022_10_13.csv"))
center_counts <- read_csv(here("DataProcessed/rna_seq/differential_expression/full_analysis/center_ruv_normcounts_2022_10_13.csv"))
ruv_counts <- read_csv(here("DataProcessed/rna_seq/ruv/RUV_k2_norm_counts_2022_10_13.csv"))
gene_annotations <- read_tsv(here("DataRaw/RNA_Seq/gencode_annotations_choo.txt"))

gene_annotations_ruv <- gene_annotations[gene_annotations$ENSG %in% ruv_counts$gene,]

#add symbol for ruv genes
ruv_counts <- left_join(ruv_counts, gene_annotations_ruv, by = c("gene" = "ENSG"))

######################### P-Value Histograms ###############################
p_hist <- function(de_results) {
  temp_hist <- de_results %>% 
    ggplot(aes(x = pvalue))+
    geom_histogram()
  return(temp_hist)
}
#Vape and Center
vape_center_phist <- p_hist(vape_center_res) +
  labs(title = "Vape and Center",
       x  = "",
       y = "Number of Genes")
#Vape Only
vape_only_phist <- p_hist(vape_res) +
  labs(title = "Vape Only") + 
  theme(axis.title.y = element_blank())
#Center Only
center_only_phist <- p_hist(center_res) + 
  labs(title = "Center Only",
       x = "") +
  theme(axis.title.y = element_blank())

ggarrange(vape_center_phist, vape_only_phist, center_only_phist, nrow = 1)
```

From the p-value histograms above, there is no evidence of poor model fit. Table 3 shows shows both the total number of significant genes for each model (FDR < 0.05) followed by the number of genes that were significant in only that model (FDR < 0.05). The *Center Only* model had 0 unique significant genes. 


*Table 3: Summary of Significant LRT Results (Full Model the same across all tests)*
```{r}
######################### Results Table ###############################
#join all results
all_results <- left_join(vape_center_res, vape_res, by = "ensg", suffix = c(".vape_center", ".vape_only"))

all_results <- left_join(all_results, center_res, by = "ensg", suffix = c("",".center_only"))

all_pvals <- all_results %>% 
  select(ensg, padj.vape_center, padj.vape_only, padj) %>% 
  dplyr::rename(gene = ensg,
                vape_center = padj.vape_center,
                vape_only = padj.vape_only,
                center_only = padj)

all_pvals <- all_pvals %>% 
  dplyr::mutate(vape_center_sig = if_else(vape_center < 0.05 & vape_only > 0.05 & center_only > 0.05, T,F),
                vape_only_sig = if_else(vape_center > 0.05 & vape_only < 0.05 & center_only > 0.05, T,F),
                center_only_sig = if_else(vape_center > 0.05 & vape_only > 0.05 & center_only < 0.05, T,F))


# function 
format_big <- function(x) {
  formatC(x, digits = 0, format = "f", big.mark = ",")
}

sig_gene_count <- function(de_res) {
  sum(de_res$padj < 0.05)
}

vape_center_sig <- as.numeric(sum(all_pvals$vape_center_sig))
vape_only_sig <- as.numeric(sum(all_pvals$vape_only_sig))
center_only_sig <- as.numeric(sum(all_pvals$center_only_sig))

sig_genes_tab <- tibble(Model = c("Vape and Center", "Vape Only", "Center Only"),
                        "Total Significant Genes (FDR < 0.05)" = c(format_big(sig_gene_count(vape_center_res)), 
                                                           format_big(sig_gene_count(vape_res)),
                                                           sig_gene_count(center_res)),
                        "Unique Significant Genes (FDR < 0.05)" = c(format_big(vape_center_sig), 
                                                           format_big(vape_only_sig),
                                                           center_only_sig)) %>% 
  kbl(digits = 3, booktabs = TRUE) %>% 
  kable_styling(position = "center", latex_options = "striped")
sig_genes_tab
```

To additionally check model results, Figures 5 - 7 box plots compare $\log_{10}$ transformed expression of the top 4 genes for each model as indicated by the smallest FDR. 

### Figures 5: $\log_{10}$(RUV-Normalized Expression) of top genes (Vape Status and Center)
```{r fig.width=10, fig.height=8}
######################### Top Gene Boxplots Function ###############################

#Set color pallettes to match
center_colors <- brewer.pal(10,"Paired")[4:6]
vape_colors <- brewer.pal(3,"Set2")
#Write a function
tcount_boxplot <- function(tcounts) {
  top_4 <- tcounts %>% 
    filter(symbol %in% unique(tcounts$symbol)[1:4])
  
  #Stratified by Vape and Center
  vape_center_box <- top_4 %>% 
    ggplot(aes(vape_6mo_lab, expression, group = interaction(vape_6mo_lab, recruitment_center))) + 
    geom_boxplot(aes(fill = recruitment_center)) +
    geom_text(label = if_else(top_4$Row.names == "Sample12", top_4$Row.names, ""), col = "Black") +
    facet_wrap(~symbol, scales="free_y") + 
    labs(x="", 
         y="Expression (log-normalized counts)", 
         fill="Center") +
    scale_fill_manual(values = center_colors)
  
  #Stratified by vape status only
  vape_box <- top_4 %>%  
    ggplot(aes(vape_6mo_lab, expression, fill= vape_6mo_lab)) + 
    geom_boxplot() + 
    geom_text(label = if_else(top_4$Row.names == "Sample12", top_4$Row.names, ""), col = "Black") +
    labs(x="Vape Status (6 mo)", 
         y="Expression (log-normalized counts)", 
         fill = "Vape Status (6 mo)") +
    scale_fill_manual(values = vape_colors) +
    facet_wrap(~symbol, scales="free_y") +
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          legend.position = "bottom")
  
  #Stratified by center only
  center_box <- top_4 %>% 
    ggplot(aes(recruitment_center, expression, fill=recruitment_center)) + 
    geom_boxplot() +
    geom_text(label = if_else(top_4$Row.names == "Sample12", top_4$Row.names, ""), 
              col = "Black") +
    facet_wrap(~symbol) + 
    labs(x="Recruitment Center", 
         y="Expression (log-normalized counts)", 
         fill="Center") +
    scale_fill_manual(values = center_colors) + 
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          legend.position = "bottom") 
  
  return(list(vape_center_box, vape_box, center_box))
    
}

######################### Top Gene Boxplots ###############################
#Vape and Center
vape_center_plots <- tcount_boxplot(tcounts = vape_center_counts)

#Vape only
vape_plots <- tcount_boxplot(tcounts = vape_counts)

#Center Only
center_plots <- tcount_boxplot(tcounts = center_counts)

ggarrange(plotlist = list(vape_center_plots[[2]], vape_center_plots[[3]]), nrow = 1)

```


Looking at the vape status boxplots, there is an obvious visual difference in the mdedian values. For recruitment center plots, the median lines are much closer together. With the inclusion of the RUVr factors, sample 12 no longer is a suspected outlier.

### Figure 6: $log_{10}$(RUVr-Normalized Expression) of top genes (Vape Status Only)
```{r, fig.width= 10, fig.height= 8}
ggarrange(plotlist = list(vape_plots[[2]], vape_plots[[3]]), nrow = 1)
```
  The plots for the *Vape Status Only* model has similar trends to the *Vape and Center* model. There do not appear to be center effects that are unaccounted for in comparison to the *Vape and Center* model. 
  
### Figure 7: $\log_{10}$(RUVr-Normalized Expression) of top genes (Center Only)
```{r fig.width= 10, fig.height= 8}  
ggarrange(plotlist = list(center_plots[[2]], center_plots[[3]]), nrow = 1)
```

The inclusion of the RUV factors in the selected models seem to correct the outlying nature of Sample 12 like the PCA and RLE plots (Figures 4 and 5) suggested.

# Discussion/Conclusion

The results from DESeq2 suggest there are differences in expression for roughly 7,000 genes after filtering of low-quality genes and normalization. This is a surprisingly *large* number. Histograms of $log_2(Fold \ Change)$ estimates suggest that *many* of the estimates are close to 0. This may be a reflection of a small sample size and high variability in the data. In further analyses, a cutoff-value for $log_2(Fold \ Change)$ will be explored in order to capture the most robust changes in gene expression.

RUVr performed well in correcting the outlying nature of Sample 12, but sensitivity analyses will still need to be conducted to ensure that inclusion of the sample is not adding significant bias to the data. There is additional concern for *recruitment center* as a potential source of confounding or multicollinearity since almost all subjects are recruited from Pueblo. A separate sensitivity analysis will be conducted for this covariate to assess modeling bias.  

In the top genes (determined by adjusted p-value), there were no obvious visual differences between the model that tested both vape status *and* recruitment center and the model that tested only vape status (the model of interest). The top genes are consistent between the models as well. This is a desirable trend and suggests center is not adding greatly to the model above what vape status accounts for already.



<!-- footer -->

------------------------------------------------------------------------

```{r, echo=FALSE, out.width='70%', fig.show='hold'}
knitr::include_graphics(logo)
```
