---
title: Metadata Exploratory Report
author: |
 | Project: `r CIDAtools::ProjectName()`
 |
 | Analyst: `r CIDAtools::ProjectAnalyst()`
 |
 | Investigator(s): `r CIDAtools::ProjectPI()`
 |
 | Report generated: `r paste(format(Sys.Date(), '%B %d, %Y'))`
output:
  html_document:
    highlight: espresso
    number_sections: no
    theme: yeti
    toc: yes
    toc_depth: 2
    toc_float: yes
    code_folding: hide
---

```{r, echo=FALSE, out.width='70%', fig.show='hold'}
logo <- system.file("figures", "sph_cida_wm_blk.png", package="CIDAtools")
knitr::include_graphics(logo)
```

------------------------------------------------------------------------

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(tidyverse)
library(knitr)
library(kableExtra)
library(arsenal)
library(here)
library(readr)
library(coin)
library(latex2exp)
library(ggpubr)
library(DESeq2)
library(RUVSeq)
library(janitor)
library(dendextend)
library(WGCNA)
```

```{r import data, include=FALSE}
#import data
tab1_dat <- read_csv(here("DataProcessed/metadata_cleaning/table1_clean_data_2022_03_02.csv"))

tab1_dat$grade <- as_factor(tab1_dat$grade)

#filter out 1 NA for vape status
tab1_dat <- tab1_dat %>% 
  filter(!is.na(vape_6mo_lab))

```

# Background

The data presented in this report are part of a study aimed to assess differential gene expression and methylation in vaping versus non-vaping LatinX youths in Pueblo and Denver, CO. Pulmonary function data were also obtained in order to better understand the impacts of vape use on pulmonary function. To assess differential gene expression and methylation, naso-epithelial swabs were obtained from each participating subject. Pulmonary function is assessed using PFTs (Pulmonary Function Tests) and Impulse Oscillometry (IOS).

# Study Population

This data set consists of samples taken from 51 people ages 12-17 yrs from the Pueblo, Denver, and Aurora, CO areas. Vape Status (did you vape in the last 6 months?) and ethnicity are self-reported.

# Methods

*All analyses performed using `r R.version$version.string`*

## Clinical Data Pocessing

### *Vape Status*

Subjects are dichotomized to those that used a vaping device in the last 6 months and those who have not based on the variables *ever_vape*, *vape_days*, and *last_vape*. This variable will be referred to as *Vape Status* throughout this report. One participant (SID = 111) reported that they had used a vaping device 5 out of the last 30 days, but did not respond to *last_vape*. They were labeled as "NA" in previous analyses.

### *Sex*

Biological sex will need to be identified using available genomic data.

### *Geographic Location*

Subjects' geographic location, *city*, was grouped into the new broader variable *recruiting_center* which encompasses the broader geographic region where they live.

### *Lung Function and IOS*

Measures of lung function and IOS were visually inspected for normality using histograms.

## Gene-Count Processing

*Annotation: Ensembl annotation for GrCH 39 ver. 37*\
*Differential Expression Analysis: DESeq2 `r package.version("DESeq2")`*\
*Removal of Unwanted Variance: RUVSeq `r package.version("RUVSeq")`*

*Gene Filtering Parameters*\
This analysis will conduct a comparison of various gene-filtering parameters presented in previous analyses and in the current literature to select parameters best-suited for this study.

*Normalization*\
The following analyses used the function RUVr from the R package RUVSeq. RUVr uses the deviance residuals from a first pass negative binomial GLM to perform a factor analysis which corrects for unwanted technical effects. The first-pass model formula is presented below :

$$raw \ read \ count \sim \beta_0 * vape \ status \ + \beta_1 *male \ + \beta_2 * latinx$$

RUVr will be performed with k = 1 through k = 5 factors and the best cutoff for factor analysis will be determined visually using RLE plots and dendrograms for each level k. Previous analyses used R package DESeq2 to fit the first-pass GLM. This analysis will use edgeR due to its reference in the literature for the RUVr procedure mentioned above.

*Transformations*\
A variance stabilizing transformation (VST) was applied in previous analyses. This analysis will use the same transformation, but will apply the transformation only after normalization.

# Results

After removing participants with missing values for *vape status*, we are left with **n = 50** subjects. Previous analyses showed **n = 12** participants had vaped in the last 6 months. This analysis will use **n = 13** participants who had vaped in the last 6 months. The lung function variable *`r colnames(tab1_dat[,12])`* and *`r colnames(tab1_dat[,13])`* reported 22 missing values. IOS measures *`r colnames(tab1_dat[,14])`* and *`r colnames(tab1_dat[,15])`* reported 1 and 6 missing values, respectively.

*Table 1: Clinical Data*

```{r Table 1, message=FALSE, results='asis', echo=TRUE}

#create table 1
vape_6mo_table1 <- tab1_dat %>% 
  tableby(vape_6mo_lab ~ gender_lab + age + recruitment_center + 
            latino_lab + fev1 + fev1_fvc + 
            r5 + x20, data = ., digits = 1, test = FALSE )

#Fix Labels
arsenal::labels(vape_6mo_table1) <- c(vape_6mo_lab = "Vaped in last 6 months", 
                             age = "Age (yrs)", gender_lab = "Gender", 
                             recruitment_center = "Recruitment Center", 
                             latino_lab = "Ethnicity", fev1 = 'FEV1',
                             fev1_fvc = "FEV1/FVC (%)", r5 = 'R5', x20 = 'X20')

#Print Tables
summary(vape_6mo_table1, pfootnote = T)
```

```{r figures, echo=TRUE, message=FALSE, warning=FALSE}
#Fix Gender Bar Plot
gender_plot_trial <- tab1_dat %>% 
  group_by(gender_lab, vape_6mo_lab) %>% 
  summarise(N = n())

gender_plot_trial[nrow(gender_plot_trial) + 1,] <- NA

gender_plot_trial[6,] <- list("Non-Binary", "Vaped in Last 6 Months", 0)


#By Recruitment gender (n = 50)
bar_gender <- gender_plot_trial %>% 
  ggplot(aes(x = gender_lab, y = N, fill = vape_6mo_lab)) + 
  geom_bar(stat = 'identity', position = 'dodge') +
  labs(x = "Gender", y = "Count", fill = "Vape Status") + 
  ggtitle("Gender")

#By Latino(n = 50)
bar_latino <- tab1_dat %>% 
  ggplot(aes(x = latino_lab, fill = vape_6mo_lab))+
  geom_bar(position = "dodge")+
  labs(x = "Ethnicity", y = "Count", fill = "Vape Status") +
  ggtitle("LatinX")

#Fix Recruitment Center Bar Plot
center_plot_trial <- tab1_dat %>% 
  group_by(recruitment_center, vape_6mo_lab) %>% 
  summarise(N = n())

center_plot_trial[nrow(center_plot_trial) + 1,] <- NA

center_plot_trial[6,] <- list("Aurora", "Vaped in Last 6 Months", 0)

#By Recruitment Center
recruitment_center_hist <- center_plot_trial %>% 
  ggplot(aes(x = recruitment_center, y = N, fill = vape_6mo_lab)) + 
  geom_bar(stat = 'identity', position = 'dodge') +
  labs(x = "Center", y = "Count", fill = "Vape Status") + 
  ggtitle("Center")


#By FEV1/FVC Continuous
hist_fev1_fvc <- tab1_dat %>%
  ggplot(aes(x = fev1_fvc, fill = vape_6mo_lab))+
  geom_histogram(binwidth = 0.1, breaks = seq(0.5,1,0.1), position = "dodge")+
  labs(x = TeX("\\frac{FEV1}{FVC}"), y = "Count", fill = "Vape Status:") +
  xlim(0.5,1) + 
  ggtitle("FEV1/FVC")

#By FEV1/FVC Box
box_fev1_fvc <- tab1_dat %>%
  ggplot(aes(x = vape_6mo_lab, y = fev1_fvc, fill = vape_6mo_lab)) +
  geom_boxplot(show.legend = F) +
  labs(x = "Vape Status", y = TeX("\\frac{FEV1}{FVC}")) +
  ggtitle(TeX("\\frac{FEV1}{FVC} \ by Vape Status (n = 28)"))

#By R5(hist)
r5_hist <- tab1_dat %>%
  ggplot(aes(x = r5, fill = vape_6mo_lab)) +
  geom_histogram(binwidth = 1, breaks = seq(1,8,1), position = 'dodge') +
  labs(x = "R5", y = "Count", fill = "Vape Status") +
  xlim(1,8) +
  ggtitle("R5")

#by R5(box)
r5_box <- tab1_dat %>%
  ggplot(aes(x = vape_6mo_lab, y = r5, fill = vape_6mo_lab)) +
  geom_boxplot(show.legend = F) +
  labs(x = "Vape Status", y = "R5") +
  ggtitle("R5 by Vape Status (n = 49)")

#by X20(box)
x20_box <- tab1_dat %>%
  ggplot(aes(x = vape_6mo_lab, y = x20, fill = vape_6mo_lab)) +
  geom_boxplot(show.legend = F) +
  labs(x = "Vape Status", y = "X20") +
  ggtitle("X20 by Vape Status (n = 44)")

#by x20(his)
x20_hist <- tab1_dat %>%
  ggplot(aes(x = x20, fill = vape_6mo_lab)) +
  geom_histogram(bins = 10, breaks = seq(-2,3,1), position = "dodge") +
  labs(x = 'X20', y = "Count", fill = 'Vape Status') +
  ggtitle("X20")

```

*Figure 1: Population Demographics*

Figure 1 visualizes the demographic information presented in Table 1. The majority of vaping subjects were recruited in Pueblo (92%) and identified as LatinX (85%). 62% of vapers identified as male.

```{r Demographic Figure, include=T, fig.height=6, fig.width=9}
#Demographic Figures
ggarrange(recruitment_center_hist, ggarrange(bar_gender, bar_latino, ncol = 2, nrow = 1, legend = "none"), nrow = 2, common.legend = T, legend = "bottom")
```

*Figure 2: Pulmonary Function*

Figure 2 is a visualization of the pulmonary function ($\frac{FEV1}{FVC}$) and IOS (*R5* and *X20*) variables. $\frac{FEV1}{FVC}$ was only completed by **n = 22** individuals from the study population. *R5* and *X20* represent **n = 49** and **n = 44** individuals, respectively.

```{r Respiratory Figure, echo=T, fig.height=6, fig.width=8, message=FALSE, warning=FALSE, include=TRUE}
#Respiratory Figures
ggarrange(hist_fev1_fvc, ggarrange(r5_box, x20_box, ncol = 2, nrow = 1, legend = "none"), nrow = 2, common.legend = T, legend = "bottom")
```

*Table 2: Comparison of Gene Filtering Parameters*

The table below compares gene-filtering parameters from previous analyses to a parameter presented by the creators of the RUVseq package.

```{r gene_compare_table, echo=TRUE, message=FALSE, warning=FALSE}
source(here("Code/exploratory_report/01a_gene_filter_comparison.R"), local = knitr::knit_global())

filter_compar
```

*Figure 3: Genes removed for read-count cutoff values*
The figure below is used to visually assess how many genes are removed across cutoff values for the range of reads for each gene across the 49 samples after filtering out genes that have 0 read counts in 75% or more of the samples. The red line represents the number of genes (14678) removed at the cutoff range of 100 (the value used in previous analyses).

```{r}
range_cutoff
```


*Figure 4: Relative Loge Expression and Principal Component Analysis Prior to RUV*

The following figure shows the Relative Log Expression (RLE) and Principal Component Analysis (PCA) of read counts for each sample without the use of RUVr or any other transformation technique. Sample23 (SID = 144) has been removed due to missing vape status. 

```{r ruvr, message=FALSE, warning=FALSE, fig.height=8, fig.width= 10}
source(here("Code/exploratory_report/01_gene_filter_and_ruv.R"), local = knitr::knit_global())

#Look at raw sample
library(RColorBrewer)
colors <- brewer.pal(3,"Set2")
par(mfrow = c(2,1))
plotRLE(ruv_prep, outline=FALSE, ylim=c(-4, 4), col = colors[metadata_joined$vape_6mo_lab])
plotPCA(ruv_prep, col = colors[metadata_joined$vape_6mo_lab])
```

*Figure 5: RLE and PCA Excluding Outlier Samples*
It appears that Sample 12 (SID = 102) may be an outlier. Below are the RLE and PCA plots with the sample removed. 

```{r, message=FALSE, warning=FALSE,fig.height=8, fig.width= 10}
par(mfrow = c(2,1))
plotRLE(ruv_ready, outline=FALSE, ylim=c(-4, 4), col = colors[metadata_joined$vape_6mo_lab] )
plotPCA(ruv_ready, col = colors[metadata_joined$vape_6mo_lab])
```

Both the RLE and PCA plots appear to improve when Sample 12 is excluded. All following analyses will exclude the sample. 

*Figure 6: Comparison of RUVr Factor Inclusion*
To visually inspect for the best cutoff for factor analysis, RUVr was run for k = 1 through k = 5 factors. The RLE plots for each are presented below.

```{r RUVr Continues, message=FALSE, warning=FALSE, fig.height=8, fig.width= 10}
#Make nicer plot
par(mfrow = c(5,1))

plotRLE(ruv_k1, outline=FALSE, ylim=c(-1,1), col = colors[metadata_joined$vape_6mo_lab], ylab = "k = 3")
abline(h = c(.5, -.5), col = c("red", "red"))

plotRLE(ruv_k2, outline=FALSE, ylim=c(-1,1), col = colors[metadata_joined$vape_6mo_lab], ylab = "k = 3")
abline(h = c(.5, -.5), col = c("red", "red"))

plotRLE(ruv_k3, outline=FALSE, ylim=c(-1,1), col = colors[metadata_joined$vape_6mo_lab], ylab = "k = 3")
abline(h = c(.5, -.5), col = c("red", "red"))

plotRLE(ruv_k4, outline=FALSE, ylim=c(-1,1), col = colors[metadata_joined$vape_6mo_lab], ylab = "k = 4")
abline(h = c(.5, -.5), col = c("red", "red"))

plotRLE(ruv_k5, outline=FALSE, ylim=c(-1,1), col = colors[metadata_joined$vape_6mo_lab], ylab = "k = 5")
abline(h = c(.5, -.5), col = c("red", "red"))

```

The dendrograms for k = 3 and k = 4 are compared below

```{r}
#Make Dendrograms
calDend <- function(x){ 
  hc <- hclust(as.dist(1-bicor(normCounts(x))))
  hc$labels <- as.factor(str_replace(hc$labels, "_", "."))
  hcd <- as.dendrogram(hc)
  return(hcd)
}

par(mfrow=c(2,1), cex = 0.5, cex.lab = 1.5, las = 3, mar=c(4.5,5,0.5,0))
k3_dend <- plot(calDend(ruv_k3), ylab = "k=3", col = colors[metadata_joined$vape_6mo_lab])
k4_dend <- plot(calDend(ruv_k4), ylab = "k=4", col = colors[metadata_joined$vape_6mo_lab])
```

# Notes


<!-- footer -->

------------------------------------------------------------------------

```{r, echo=FALSE, out.width='70%', fig.show='hold'}
knitr::include_graphics(logo)
```
