---
title: Metadata Exploratory Report
author: |
 | Project: `r CIDAtools::ProjectName()`
 |
 | Analyst: `r CIDAtools::ProjectAnalyst()`
 |
 | Investigator(s): `r CIDAtools::ProjectPI()`
 |
 | Report generated: `r paste(format(Sys.Date(), '%B %d, %Y'))`
output:
  html_document:
    highlight: espresso
    number_sections: no
    theme: yeti
    toc: yes
    toc_depth: 2
    toc_float: yes
    code_folding: hide
---

```{r, echo=FALSE, out.width='70%', fig.show='hold'}
logo <- system.file("figures", "sph_cida_wm_blk.png", package="CIDAtools")
knitr::include_graphics(logo)
```

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(tidyverse)
library(knitr)
library(kableExtra)
library(arsenal)
library(here)
library(readr)
library(coin)
library(latex2exp)
library(ggpubr)
library(DESeq2)
library(RUVSeq)
```


```{r import data, include=FALSE}
#import data
tab1_dat <- read_csv(here("DataProcessed/metadata_cleaning/table1_clean_data_2022_03_02.csv"))

tab1_dat$grade <- as_factor(tab1_dat$grade)

#filter out 1 NA for vape status
tab1_dat <- tab1_dat %>% 
  filter(!is.na(vape_6mo_lab))

```

# Background

The data presented in this report are part of a study aimed to assess differential gene expression and methylation in vaping versus non-vaping LatinX youths in Pueblo and Denver, CO. Pulmonary function data were also obtained in order to better understand the impacts of vape use on pulmonary function. To assess differential gene expression and methylation, naso-epithelial swabs were obtained from each participating subject. Pulmonary function is assessed using PFTs (Pulmonary Function Tests) and Impulse Oscillometry (IOS).

# Study Population  

This data set consists of samples taken from 51 people ages 12-17 yrs from the Pueblo, Denver, and Aurora, CO areas. Vape Status (did you vape in the last 6 months?) and ethnicity are self-reported.

# Methods
*All analyses performed using `r R.version$version.string`*  

## Clinical Data Pocessing  
### *Vape Status*  
Subjects are dichotomized to those that used a vaping device in the last 6 months and those who have not based on the variables *ever_vape*, *vape_days*, and *last_vape*. This variable will be referred to as *Vape Status* throughout this report. One participant (SID = 111) reported that they had used a vaping device 5 out of the last 30 days, but did not respond to *last_vape*. They were labeled as "NA" in previous analyses.

&nbsp;  


### *Sex*  
Biological sex will need to be identified using available genomic data. 

### *Geographic Location*  
Subjects' geographic location, *city*, was grouped into the new broader variable *recruiting_center* which encompasses the broader geographic region  where they live.  

### *Lung Function and IOS*  
Measures of lung function and IOS were visually inspected for normality using histograms.     

## Gene-Count Processing
*Annotation: Ensembl annotation for GrCH 39 ver. 37 *  
*Differential Expression Analysis: DESeq2 `r package.version("DESeq2")`*  
*Removal of Unwanted Variance: RUVSeq `r package.version("RUVSeq")`*  

&nbsp;  

  * Gene Filtering Parameters\:
    * Removed genes with count = 0 and genes with range of counts > 100
  * Normalization\:  
    * Variance Stabilizing Transformation (VST) found in code
  * Technical effects\:
    * Used RUVr which  uses factor analysis on residuals 
    * Choo used K = 4 factors (identified from elbow plot)
    
    

# Results
After removing participants with missing values for *vape status*, we are left with **n = 50** subjects.  Previous analyses showed **n = 12** participants had vaped in the last 6 months. This analysis will use **n = 13** participants who had vaped in the last 6 months. The lung function variable *`r colnames(tab1_dat[,12])`* and *`r colnames(tab1_dat[,13])`* reported 22 missing values. IOS measures *`r colnames(tab1_dat[,14])`* and *`r colnames(tab1_dat[,15])`* reported 1 and 6 missing values, respectively.

*Table 1: Clinical Data*
```{r Table 1, message=FALSE, results='asis', echo=TRUE}

#create table 1
vape_6mo_table1 <- tab1_dat %>% 
  tableby(vape_6mo_lab ~ gender_lab + age + recruitment_center + 
            latino_lab + fev1 + fev1_fvc + 
            r5 + x20, data = ., digits = 1, test = FALSE )

#Fix Labels
labels(vape_6mo_table1) <- c(vape_6mo_lab = "Vaped in last 6 months", 
                             age = "Age (yrs)", gender_lab = "Gender", 
                             recruitment_center = "Recruitment Center", 
                             latino_lab = "Ethnicity", fev1 = 'FEV1',
                             fev1_fvc = "FEV1/FVC (%)", r5 = 'R5', x20 = 'X20')

#Print Tables
summary(vape_6mo_table1, pfootnote = T)
```


```{r figures, echo=TRUE, message=FALSE, warning=FALSE}
#Fix Gender Bar Plot
gender_plot_trial <- tab1_dat %>% 
  group_by(gender_lab, vape_6mo_lab) %>% 
  summarise(N = n())

gender_plot_trial[nrow(gender_plot_trial) + 1,] <- NA

gender_plot_trial[6,] <- list("Non-Binary", "Vaped in Last 6 Months", 0)


#By Recruitment gender (n = 50)
bar_gender <- gender_plot_trial %>% 
  ggplot(aes(x = gender_lab, y = N, fill = vape_6mo_lab)) + 
  geom_bar(stat = 'identity', position = 'dodge') +
  labs(x = "Gender", y = "Count", fill = "Vape Status") + 
  ggtitle("Gender")

#By Latino(n = 50)
bar_latino <- tab1_dat %>% 
  ggplot(aes(x = latino_lab, fill = vape_6mo_lab))+
  geom_bar(position = "dodge")+
  labs(x = "Ethnicity", y = "Count", fill = "Vape Status") +
  ggtitle("LatinX")

#Fix Recruitment Center Bar Plot
center_plot_trial <- tab1_dat %>% 
  group_by(recruitment_center, vape_6mo_lab) %>% 
  summarise(N = n())

center_plot_trial

center_plot_trial[nrow(center_plot_trial) + 1,] <- NA

center_plot_trial[6,] <- list("Aurora", "Vaped in Last 6 Months", 0)

#By Recruitment Center
recruitment_center_hist <- center_plot_trial %>% 
  ggplot(aes(x = recruitment_center, y = N, fill = vape_6mo_lab)) + 
  geom_bar(stat = 'identity', position = 'dodge') +
  labs(x = "Center", y = "Count", fill = "Vape Status") + 
  ggtitle("Center")


#By FEV1/FVC Continuous
hist_fev1_fvc <- tab1_dat %>%
  ggplot(aes(x = fev1_fvc, fill = vape_6mo_lab))+
  geom_histogram(binwidth = 0.1, breaks = seq(0.5,1,0.1), position = "dodge")+
  labs(x = TeX("\\frac{FEV1}{FVC}"), y = "Count", fill = "Vape Status:") +
  xlim(0.5,1) + 
  ggtitle("FEV1/FVC")

#By FEV1/FVC Box
box_fev1_fvc <- tab1_dat %>%
  ggplot(aes(x = vape_6mo_lab, y = fev1_fvc, fill = vape_6mo_lab)) +
  geom_boxplot(show.legend = F) +
  labs(x = "Vape Status", y = TeX("\\frac{FEV1}{FVC}")) +
  ggtitle(TeX("\\frac{FEV1}{FVC} \ by Vape Status (n = 28)"))

#By R5(hist)
r5_hist <- tab1_dat %>%
  ggplot(aes(x = r5, fill = vape_6mo_lab)) +
  geom_histogram(binwidth = 1, breaks = seq(1,8,1), position = 'dodge') +
  labs(x = "R5", y = "Count", fill = "Vape Status") +
  xlim(1,8) +
  ggtitle("R5")

#by R5(box)
r5_box <- tab1_dat %>%
  ggplot(aes(x = vape_6mo_lab, y = r5, fill = vape_6mo_lab)) +
  geom_boxplot(show.legend = F) +
  labs(x = "Vape Status", y = "R5") +
  ggtitle("R5 by Vape Status (n = 49)")

#by X20(box)
x20_box <- tab1_dat %>%
  ggplot(aes(x = vape_6mo_lab, y = x20, fill = vape_6mo_lab)) +
  geom_boxplot(show.legend = F) +
  stat_summary()
  labs(x = "Vape Status", y = "X20") +
  ggtitle("X20 by Vape Status (n = 44)")

#by x20(his)
x20_hist <- tab1_dat %>%
  ggplot(aes(x = x20, fill = vape_6mo_lab)) +
  geom_histogram(bins = 10, breaks = seq(-2,3,1), position = "dodge") +
  labs(x = 'X20', y = "Count", fill = 'Vape Status') +
  ggtitle("X20")

```


## *Figure 1: Population Demographics*

Figure 1 visualizes the demographic information presented in Table 1. The majority of vaping subjects were recruited in Pueblo (92%) and identified as LatinX (85%). 62% of vapers identified as male.

```{r Demographic Figure, include=T, fig.height=6, fig.width=9}
#Demographic Figures
ggarrange(recruitment_center_hist, ggarrange(bar_gender, bar_latino, ncol = 2, nrow = 1, legend = "none"), nrow = 2, common.legend = T, legend = "bottom")
```


## *Figure 2: Pulmonary Function*

Figure 2 is a visualization of the pulmonary function ($\frac{FEV1}{FVC}$) and IOS (*R5* and *X20*) variables. $\frac{FEV1}{FVC}$ was only completed by **n = 22** individuals from the study population. *R5* and *X20* represent **n = 49** and **n = 44** individuals, respectively. 

```{r Respiratory Figure, echo=T, fig.height=6, fig.width=8, message=FALSE, warning=FALSE}
#Respiratory Figures
ggarrange(hist_fev1_fvc, ggarrange(r5_box, x20_box, ncol = 2, nrow = 1, legend = "none"), nrow = 2, common.legend = T, legend = "bottom")
```


# Notes

- We have identified one extra vaper (n = 13)

# Questions

- It looks like the 'sex' variable in the survey included an option for 'non-binary'. Does this effectively make it a 'gender' variable? It looks like Choo handled this in the past by creating a 'Male' and 'Non-male' dummy variable, but should we identify gender using RNASeq or Methylation data?

- The report from Choo was helpful. We can see that we are stepping in after several itterations of reports from Choo. Can we get your interpretation/summary of where you left off with Choo? For example, it looks like she was investigating a center effect, but we would like to know the conclusion of that investigation. Did you have any idea of where you were headed next? 
      


<!-- footer -->

---

```{r, echo=FALSE, out.width='70%', fig.show='hold'}
knitr::include_graphics(logo)
```

