---
title: Metadata Exploratory Report
author: |
 | Project: `r CIDAtools::ProjectName()`
 |
 | Analyst: `r CIDAtools::ProjectAnalyst()`
 |
 | Investigator(s): `r CIDAtools::ProjectPI()`
 |
 | Report generated: `r paste(format(Sys.Date(), '%B %d, %Y'))`
output:
  html_document:
    highlight: espresso
    number_sections: no
    theme: yeti
    toc: yes
    toc_depth: 2
    toc_float: yes
    code_folding: hide
---

```{r, echo=FALSE, out.width='70%', fig.show='hold'}
logo <- system.file("figures", "sph_cida_wm_blk.png", package="CIDAtools")
knitr::include_graphics(logo)
knitr::opts_chunk$set(warning = F, message = F, error = F, echo = T)
```

------------------------------------------------------------------------

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(RColorBrewer)
library(tidyverse)
library(knitr)
library(kableExtra)
library(arsenal)
library(here)
library(readr)
library(coin)
library(latex2exp)
library(ggpubr)
library(DESeq2)
library(RUVSeq)
library(janitor)
library(dendextend)
library(WGCNA)
library(minfi)
```

```{r import data, include=FALSE}
#import data
source(here("Code/exploratory_report/00_clean_metadata.R"), local = knitr::knit_global())

#filter out 1 NA for vape status
tab1_dat <- tab1_dat %>% 
  filter(!is.na(vape_6mo_lab))

```

# Background

The data presented in this report are part of a study aimed to assess differential gene expression and methylation in vaping versus non-vaping LatinX youths in Pueblo and Denver, CO. Pulmonary function data were also obtained in order to better understand the impacts of vape use on pulmonary function. To assess differential gene expression and methylation, naso-epithelial swabs were obtained from each participating subject. Pulmonary function is assessed using PFTs (Pulmonary Function Tests) and Impulse Oscillometry (IOS).

# Study Population

This data set consists of samples taken from 51 people ages 12-17 yrs from the Pueblo, Denver, and Aurora, CO areas. Vape Status (did you vape in the last 6 months?) and ethnicity are self-reported.

# Methods

*All analyses performed using `r R.version$version.string`*

## Clinical Data Pocessing

### *Vape Status*

Subjects are dichotomized to those that used a vaping device in the last 6 months and those who have not based on the variables *ever_vape*, *vape_days*, and *last_vape*. This variable will be referred to as *Vape Status* throughout this report. One participant (SID = 111) reported that they had used a vaping device 5 out of the last 30 days, but did not respond to *last_vape*. That participant is recorded as *Vaped* in this analysis.

### *Sex*

Biological sex has been derived from methylation data utilizing the getSex function from the R package *minfi `r package.version("minfi")`*.

### *Geographic Location*

Subjects' geographic location, *city*, was grouped into the new broader variable *recruiting_center* which encompasses the broader geographic region where they live.

### *Lung Function and IOS*

Measures of lung function and IOS were visually inspected for normality using histograms.

## Gene-Count Processing

*Annotation: Ensembl annotation for GrCH 39 ver. 37*\
*Removal of Unwanted Variance: edgeR `r package.version("edgeR")` and RUVSeq `r package.version("RUVSeq")`*\
*Differential Expression Analysis: DESeq2 `r package.version("DESeq2")`*\


*Gene Filtering Parameters*\
This analysis will conduct a comparison of various gene-filtering parameters presented in previous analyses and in the current literature to select parameters best-suited for this study.

*Normalization*\
The following analyses used the function RUVr from the R package RUVSeq. RUVr uses the deviance residuals from a first pass negative binomial GLM to perform a factor analysis which corrects for unwanted technical effects. The first-pass model formula is presented below :

$$raw \ read \ count \sim \beta_0 + \beta_1 * vape \ status \ + \beta_2 *male \ + \beta_3 * age$$

RUVr will be performed with k = 1 through k = 2 factors and the best k for factor analysis will be determined visually using an elbow plot, RLE plots, and dendrograms. Previous analyses used R package DESeq2 to fit the first-pass GLM. This analysis will use edgeR due to its reference in the literature for the RUVr procedure mentioned above.

*Transformations*\
A variance stabilizing transformation (VST) was applied in previous analyses. This analysis will use the same transformation, but will apply the transformation only after normalization.

# Results

## Clinical Data Processing Results

After removing participants with missing values for *vape status*, we are left with **n = 50** subjects. Previous analyses showed **n = 12** participants had vaped in the last 6 months. This analysis will use **n = 13** participants who had vaped in the last 6 months. The lung function variable *`r colnames(tab1_dat[,"fev1"])`* and *`r colnames(tab1_dat[,"fev1_fvc"])`* reported 22 missing values. IOS measures *`r colnames(tab1_dat[,"r5"])`* and *`r colnames(tab1_dat[,"x20"])`* reported 1 and 6 missing values, respectively.

### Table 1: Clinical Data

```{r Table 1, message=FALSE, results='asis', echo=TRUE}

#create table 1
vape_6mo_table1 <- tab1_dat %>% 
  tableby(vape_6mo_lab ~ sex_lab + age + recruitment_center + 
            latino_lab + fev1 + fev1_fvc + 
            r5 + x20, data = ., digits = 1, test = FALSE )

#Fix Labels
arsenal::labels(vape_6mo_table1) <- c(vape_6mo_lab = "Vaped in last 6 months", 
                             age = "Age (yrs)", sex_lab = "Sex", 
                             recruitment_center = "Recruitment Center", 
                             latino_lab = "Ethnicity", fev1 = 'FEV1',
                             fev1_fvc = "FEV1/FVC (%)", r5 = 'R5', x20 = 'X20')

#Print Tables
summary(vape_6mo_table1, pfootnote = T)
```

```{r figures, echo=TRUE, message=FALSE, warning=FALSE}
#Fix sex Bar Plot
sex_plot_trial <- tab1_dat %>% 
  group_by(sex_lab, vape_6mo_lab) %>% 
  summarise(N = n())


#By Recruitment sex (n = 50)
bar_sex <- sex_plot_trial %>% 
  ggplot(aes(x = sex_lab, y = N, fill = vape_6mo_lab)) + 
  geom_bar(stat = 'identity', position = 'dodge') +
  labs(y = "Count", fill = "Vape Status") + 
  ggtitle("Sex") +
  scale_fill_manual(values = brewer.pal(3, "Set2"))+
  theme(axis.title.x=element_blank())

#By Latino(n = 50)
bar_latino <- tab1_dat %>% 
  ggplot(aes(x = latino_lab, fill = vape_6mo_lab))+
  geom_bar(position = "dodge")+
  labs( y = "Count", fill = "Vape Status") +
  ggtitle("LatinX") +
  scale_fill_manual(values = brewer.pal(3, "Set2")) +
  theme(axis.title.x=element_blank())

#Fix Recruitment Center Bar Plot
center_plot_trial <- tab1_dat %>% 
  group_by(recruitment_center, vape_6mo_lab) %>% 
  summarise(N = n())

center_plot_trial[nrow(center_plot_trial) + 1,] <- NA

center_plot_trial[6,] <- list("Aurora", "Vaped in Last 6 Months", 0)

#By Recruitment Center
recruitment_center_hist <- center_plot_trial %>% 
  ggplot(aes(x = recruitment_center, y = N, fill = vape_6mo_lab)) + 
  geom_bar(stat = 'identity', position = 'dodge') +
  labs(y = "Count", fill = "Vape Status") + 
  ggtitle("Center")+
  scale_fill_manual(values = brewer.pal(3, "Set2")) +
  theme(axis.title.x=element_blank())


#By FEV1/FVC Continuous
hist_fev1_fvc <- tab1_dat %>%
  ggplot(aes(x = fev1_fvc, fill = vape_6mo_lab))+
  geom_histogram(binwidth = 0.1, breaks = seq(0.5,1,0.1), position = "dodge")+
  labs(x = TeX("\\frac{FEV1}{FVC}"), y = "Count", fill = "Vape Status:") +
  xlim(0.5,1) + 
  scale_fill_manual(values = brewer.pal(3, "Set2")) +
  ggtitle("FEV1/FVC")

#By FEV1/FVC Box
box_fev1_fvc <- tab1_dat %>%
  ggplot(aes(x = vape_6mo_lab, y = fev1_fvc, fill = vape_6mo_lab)) +
  geom_boxplot(show.legend = F) +
  labs(x = "Vape Status", y = TeX("\\frac{FEV1}{FVC}")) +
  scale_fill_manual(values = brewer.pal(3, "Set2")) +
  ggtitle(TeX("\\frac{FEV1}{FVC} \ by Vape Status (n = 28)"))

#By R5(hist)
r5_hist <- tab1_dat %>%
  ggplot(aes(x = r5, fill = vape_6mo_lab)) +
  geom_histogram(binwidth = 1, breaks = seq(1,8,1), position = 'dodge') +
  labs(x = "R5", y = "Count", fill = "Vape Status") +
  xlim(1,8) +
  scale_fill_manual(values = brewer.pal(3, "Set2")) +
  ggtitle("R5")

#by R5(box)
r5_box <- tab1_dat %>%
  ggplot(aes(x = vape_6mo_lab, y = r5, fill = vape_6mo_lab)) +
  geom_boxplot(show.legend = F) +
  labs(x = "Vape Status", y = "R5") +
  ggtitle("R5 by Vape Status (n = 49)") +
  scale_fill_manual(values = brewer.pal(3, "Set2")) +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank())

#by X20(box)
x20_box <- tab1_dat %>%
  ggplot(aes(x = vape_6mo_lab, y = x20, fill = vape_6mo_lab)) +
  geom_boxplot(show.legend = F) +
  labs(x = "Vape Status", y = "X20") +
  ggtitle("X20 by Vape Status (n = 44)") +
  scale_fill_manual(values = brewer.pal(3, "Set2")) +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank())

#by x20(his)
x20_hist <- tab1_dat %>%
  ggplot(aes(x = x20, fill = vape_6mo_lab)) +
  geom_histogram(bins = 10, breaks = seq(-2,3,1), position = "dodge") +
  labs(x = 'X20', y = "Count", fill = 'Vape Status') +
  scale_fill_manual(values = brewer.pal(3, "Set2")) +
  ggtitle("X20")

```

### Figure 1: Population Demographics

Figure 1 visualizes the demographic information presented in Table 1. The majority of vaping subjects were recruited in Pueblo (92%) and identified as LatinX (85%). 62% of subjects were male as confirmed from the available methylation data.

```{r Demographic Figure, include=T, fig.height=6, fig.width=9}
#Demographic Figures
ggarrange(recruitment_center_hist, ggarrange(bar_sex, bar_latino, ncol = 2, nrow = 1, legend = "none"), nrow = 2, common.legend = T, legend = "bottom")
```

### Figure 2: Pulmonary Function

Figure 2 is a visualization of the pulmonary function ($\frac{FEV1}{FVC}$) and IOS (*R5* and *X20*) variables. $\frac{FEV1}{FVC}$ was only completed by **n = 22** individuals from the study population. *R5* and *X20* represent **n = 49** and **n = 44** individuals, respectively.

```{r Respiratory Figure, echo=T, fig.height=6, fig.width=8, message=FALSE, warning=FALSE, include=TRUE}
#Respiratory Figures
ggarrange(hist_fev1_fvc, ggarrange(r5_box, x20_box, ncol = 2, nrow = 1, legend = "none"), nrow = 2, common.legend = T, legend = "bottom")
```

  
### Figure 3: Correlation Matrix  
Figure 3 identifies correlation patterns in the clinical variables of interest. To test for correlation, a contingency table was made for each comparison of variables. because some contingency tables contained cells with <5 subjects, a Fisher's Exact test was used for all tests. 
```{r}
source(here("Code/exploratory_report/01_variable_correlation.R"), local = knitr::knit_global())

correlation_matrix_plot
```

From the correlation matrix, it appears that there may be significant correlation between vape status and recruitment center. A separate t-test was run to determine independence between vape status and age.

```{r}
age_report_tab
```


## Gene-Count Processing Results  

After merging clinical metadata with gene-count data, there were **n = 47** samples present. From the clinical metadata, SID = 105, 137, and 103 are missing genetic data. Sample 23 (SID = 144) was also removed due to missing vape status.  


### Table 2: Comparison of Gene Filtering Parameters

The table below compares gene-filtering parameters from previous analyses to a parameter presented by the creators of the RUVseq package.

```{r gene_compare_table, echo=TRUE, message=FALSE, warning=FALSE}
source(here("Code/exploratory_report/02_gene_filter_comparison.R"), local = knitr::knit_global())

filter_compar
```


### Figure 4: Genes removed for read-count cutoff values  
The figure below is used to visually assess how many genes are removed across cutoff values for the range of reads for each gene across the 47 samples after filtering out genes that have 0 read counts in 75% or more of the samples. The red point represents the number of genes (14,645) removed at the cutoff range of 100 (the value used in previous analyses).

```{r}
range_cutoff
```

After reviewing the comparison of filters, it appears that filtering parameters presented by the creators of RUVSeq may be overly conservative for this application. Analyses will proceed using the same filters as previous analyses (filters 1 & 2 from table 2). The first filter will remove genes with 0 reads in more than 75% of samples, and the second will remove genes with low variation (range of reads < 100) that may be considered "house-keeping" genes. This analysis will include a total of **n = 16,860** genes. 

### Figure 5: Relative Loge Expression and Principal Component Analysis Prior to RUV

The following figure shows the Relative Log Expression (RLE) and Principal Component Analysis (PCA) of read counts for each sample without any prior transformations.

```{r, message=FALSE, warning=FALSE, fig.height=8, fig.width= 10}
source(here("Code/exploratory_report/03_gene_filter_and_ruv_edgeR.R"), local = knitr::knit_global())

#Look at raw sample
colors <- brewer.pal(3,"Set2")
par(mfrow = c(2,1))

plotRLE(ruv_with_12, outline=FALSE, ylim=c(-4, 4), col = colors[metadata_joined$vape_6mo_lab])
legend(x = "bottomleft", legend = c("Vaped", "Did Not Vape"), fill = c(colors[2], colors[1]))

plotPCA(ruv_with_12, col = colors[metadata_joined$vape_6mo_lab])
legend(x = "bottomleft", legend = c("Vaped", "Did Not Vape"), fill = c(colors[2], colors[1]))
```

### Figure 6: RLE and PCA Excluding Outlier Samples 
It appears that Sample 12 (SID = 102) may be an outlier. Below are the RLE and PCA plots with the sample removed. 

```{r, message=FALSE, warning=FALSE,fig.height=8, fig.width= 10}
par(mfrow = c(2,1))
plotRLE(ruv_no12, outline=FALSE, ylim=c(-4, 4), col = colors[metadata_joined_no12$vape_6mo_lab] )
legend(x = "bottomright", legend = c("Vaped", "Did Not Vape"), fill = c(colors[2], colors[1]))

plotPCA(ruv_no12, col = colors[metadata_joined_no12$vape_6mo_lab])
legend(x = "bottomright", legend = c("Vaped", "Did Not Vape"), fill = c(colors[2], colors[1]))
```

To further assess the removal of sample 12, the following is an Elbow Plot which displays the reduction in within-cluster sums of squares when increasing the number of clusters (k) both with and without sample 12. 

### Figure 7: Elbow Plot Comparison of RUVr Factor Inclusion
```{r}
elbow_plot
```

To visually inspect for the best cutoff for factor analysis, RUVr was run for k = 1 and k = 2 factors both with and without Sample 12 included in the dataset. The RLE and PCA plots for each are presented below.
  
### Figure 8: RLE Comparison of RUVr Factor Inclusion (Sample 12 in dataset) 
```{r RUVr Continues, message=FALSE, warning=FALSE, fig.height=7, fig.width= 10}
#Make nicer plot
par(mfrow=c(2,1), cex = .75, cex.lab = 1.5, mar=c(2,5,0.5,0))


plotRLE(ruv_k1, outline=FALSE, ylim=c(-2,2), col = colors[metadata_joined$vape_6mo_lab], ylab = "k = 1", main = "With Sample 12")
abline(h = c(.5, -.5), col = c("red", "red"))
legend(x = "bottomright", legend = c("Vaped", "Did Not Vape"), fill = c(colors[2], colors[1]))

plotRLE(ruv_k2, outline=FALSE, ylim=c(-2,2), col = colors[metadata_joined$vape_6mo_lab], ylab = "k = 2")
abline(h = c(.5, -.5), col = c("red", "red"))
legend(x = "bottomright", legend = c("Vaped", "Did Not Vape"), fill = c(colors[2], colors[1]))

```
As was discussed previously, it appears that Sample 12 may be an outlier. Below are the RLE plots after RUVr was conducted with k = 1 and k = 2 with Sample 12 removed from the dataset.  
  
### Figure 9: RLE Comparison of RUVr Factor Inclusion (Sample 12 removed from dataset) 
```{r RUVr, message=FALSE, warning=FALSE, fig.height=7, fig.width= 10}
#Make nicer plot
par(mfrow=c(2,1), cex = .75, cex.lab = 1.5, mar=c(2,5,0.5,0))


plotRLE(ruv_k1_no12, outline=FALSE, ylim=c(-2,2), col = colors[metadata_joined_no12$vape_6mo_lab], ylab = "k = 1")
abline(h = c(.5, -.5), col = c("red", "red"))
legend(x = "bottomright", legend = c("Vaped", "Did Not Vape"), fill = c(colors[2], colors[1]))

plotRLE(ruv_k2_no12, outline=FALSE, ylim=c(-2,2), col = colors[metadata_joined_no12$vape_6mo_lab], ylab = "k = 2")
abline(h = c(.5, -.5), col = c("red", "red"))
legend(x = "bottomright", legend = c("Vaped", "Did Not Vape"), fill = c(colors[2], colors[1]))

```
To help additionally compare, the following are side-by-side PCA plots comparing samples with and without sample 12 removed before RUVr, after RUVr with k = 1, and after RUVr with k = 2.  

### Figure 10: PCA Comparison of RUVr Factor Inclusion
```{r message=FALSE, warning=FALSE, fig.height=12, fig.width= 12}
#Make nicer plot
par(mfrow=c(3,2), cex = .75)

plotPCA(ruv_with_12, col = colors[metadata_joined$vape_6mo_lab], main = "With Sample 12", xlim = c(-0.7, 0.7))
legend(x = "topright", legend = c("Vaped", "Did Not Vape"), fill = c(colors[2], colors[1]))

plotPCA(ruv_no12, col = colors[metadata_joined_no12$vape_6mo_lab], main = "Sample 12 Removed", xlim = c(-0.7, 0.7))
legend(x = "topright", legend = c("Vaped", "Did Not Vape"), fill = c(colors[2], colors[1]))


plotPCA(ruv_k1, col = colors[metadata_joined$vape_6mo_lab], main = "With Sample 12 (k = 1)", xlim = c(-0.7, 0.7))


plotPCA(ruv_k1_no12, col = colors[metadata_joined_no12$vape_6mo_lab], main = "Sample 12 Removed (k = 1)", xlim = c(-0.7, 0.7))


plotPCA(ruv_k2, col = colors[metadata_joined$vape_6mo_lab], main = "With Sample 12 (k = 2)", xlim = c(-0.7, 0.7))


plotPCA(ruv_k2_no12, col = colors[metadata_joined_no12$vape_6mo_lab], main = "Sample 12 Removed (k = 2)", xlim = c(-0.7, 0.7))

```



The dendrograms for k = 1 and k = 2 are compared below with and without Sample 12 included.  

### Figure 11: Dendrogram Comparison of RUVr Factor Inclusion (k = 1 and k = 2) (Sample 12 in dataset)
```{r, fig.height= 6, fig.width=10}
#Make Dendrograms
calDend <- function(x){ 
  hc <- hclust(as.dist(1-cor(normCounts(x))))
  hc$labels <- as.factor(str_replace(hc$labels, "_", "."))
  hcd <- as.dendrogram(hc)
  return(hcd)
}
#Make color palattes
vape_status_colors <- colors
sex_colors <- c("#00AFBB", "#E7B800")
center_colors <- brewer.pal(10,"Paired")[4:6]

#assign color for each variable
#vape status
vape_status_dend_col <- vape_status_colors[metadata_joined$vape_6mo_lab]
sex_dend_col <- sex_colors[metadata_joined$sex_lab]
center_dend_colors <- center_colors[as.factor(metadata_joined$recruitment_center)]


par(mar = c(12,4,1,1))
plot(calDend(ruv_k1), ylab = "k=1", ylim = c(0,0.6))
colored_bars(cbind(sex_dend_col, center_dend_colors, vape_status_dend_col), calDend(ruv_k1), rowLabels = c("Sex", "Recruitment Center", "Vape Status"))
legend(x = "topleft", legend = c("Vaped", "Did not Vape"), fill = c(colors[2], colors[1]), title = "Vape Status")
legend(x = "topright", legend = c("Male", "Female"), fill = c(sex_colors[2], sex_colors[1]), title = "Sex")
legend(x = "top", legend = c("Pueblo", "Denver", "Aurora"), fill = c(center_colors[3],center_colors[2], center_colors[1]), title = "Center")


par(mar = c(12,4,1,1))
plot(calDend(ruv_k2), ylab = "k=2", ylim = c(0,0.6))
colored_bars(cbind(sex_dend_col, center_dend_colors, vape_status_dend_col), calDend(ruv_k2), rowLabels = c("Sex", "Recruitment Center", "Vape Status")) 
legend(x = "topleft", legend = c("Vaped", "Did not Vape"), fill = c(colors[2], colors[1]), title = "Vape Status")
legend(x = "topright", legend = c("Male", "Female"), fill = c(sex_colors[2], sex_colors[1]), title = "Sex")
legend(x = "top", legend = c("Pueblo", "Denver", "Aurora"), fill = c(center_colors[3],center_colors[2], center_colors[1]), title = "Center")  
```

### Figure 12: Dendrogram Comparison of RUVr Factor Inclusion (k = 1 and k = 2) (Sample 12 removed)
```{r fig.height= 6, fig.width=10}
#assign color for each variable
#vape status
vape_status_dend_col <- vape_status_colors[metadata_joined_no12$vape_6mo_lab]
sex_dend_col <- sex_colors[metadata_joined_no12$sex_lab]
center_dend_colors <- center_colors[as.factor(metadata_joined_no12$recruitment_center)]


par(mar = c(12,4,1,1))
plot(calDend(ruv_k1_no12), ylab = "k=1", ylim = c(0,0.6))
colored_bars(cbind(sex_dend_col, center_dend_colors, vape_status_dend_col), calDend(ruv_k1_no12), rowLabels = c("Sex", "Recruitment Center", "Vape Status"))
legend(x = "topleft", legend = c("Vaped", "Did not Vape"), fill = c(colors[2], colors[1]), title = "Vape Status")
legend(x = "topright", legend = c("Male", "Female"), fill = c(sex_colors[2], sex_colors[1]), title = "Sex")
legend(x = "top", legend = c("Pueblo", "Denver", "Aurora"), fill = c(center_colors[3],center_colors[2], center_colors[1]), title = "Center")


par(mar = c(12,4,1,1))
plot(calDend(ruv_k2_no12), ylab = "k=2", ylim = c(0,0.6))
colored_bars(cbind(sex_dend_col, center_dend_colors, vape_status_dend_col), calDend(ruv_k2_no12), rowLabels = c("Sex", "Recruitment Center", "Vape Status")) 
legend(x = "topleft", legend = c("Vaped", "Did not Vape"), fill = c(colors[2], colors[1]), title = "Vape Status")
legend(x = "topright", legend = c("Male", "Female"), fill = c(sex_colors[2], sex_colors[1]), title = "Sex")
legend(x = "top", legend = c("Pueblo", "Denver", "Aurora"), fill = c(center_colors[3],center_colors[2], center_colors[1]), title = "Center")  
```

Using the figures above (with special attention towards Figure 10), It seems appropriate to **keep** sample 12 and use k = 2 factors for RUVr normalization. 

# Notes

- Between this report and it's previous version ("teen_vape_exploratory_report_2022_04_15"), the variable LatinX was included in the model as opposed to the variable Age. This report corrects that mistake. Visually, it seems to make little differences in the figures used to select K.  

<!-- footer -->

------------------------------------------------------------------------

```{r, echo=FALSE, out.width='70%', fig.show='hold'}
knitr::include_graphics(logo)
```
